<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Eve Zeyl Fiskebeck">
<meta name="dcterms.date" content="2024-02-05">

<title>SEP H5N5 HA Phylogenetic tree</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="SEP_H5N5_HA_phylogenetic_tree_files/libs/clipboard/clipboard.min.js"></script>
<script src="SEP_H5N5_HA_phylogenetic_tree_files/libs/quarto-html/quarto.js"></script>
<script src="SEP_H5N5_HA_phylogenetic_tree_files/libs/quarto-html/popper.min.js"></script>
<script src="SEP_H5N5_HA_phylogenetic_tree_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="SEP_H5N5_HA_phylogenetic_tree_files/libs/quarto-html/anchor.min.js"></script>
<link href="SEP_H5N5_HA_phylogenetic_tree_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="SEP_H5N5_HA_phylogenetic_tree_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="SEP_H5N5_HA_phylogenetic_tree_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="SEP_H5N5_HA_phylogenetic_tree_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="SEP_H5N5_HA_phylogenetic_tree_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">SEP H5N5 HA Phylogenetic tree</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Eve Zeyl Fiskebeck </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 5, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="setup" class="level1">
<h1>Setup</h1>
<ul>
<li>sourcing scripts and functions</li>
</ul>
<p>Note: if you are modifying the script and want to continue to work after, you can load a previous environment (don’t forget to save it at the end of your work).</p>
<p>Loads the environment (to continue previous work)</p>
</section>
<section id="preparation-of-metadata-specific-to-the-tree" class="level1">
<h1>Preparation of metadata specific to the tree</h1>
<p>This is the metadata that we combined to do all the trees, and where the host family has been added manually.</p>
<p>We select only the columns that we will use, to reduce complexity</p>
<p>Note in this script we do not have duplicate sequences, all tips represent themselves</p>
</section>
<section id="information" class="level1">
<h1>Information:</h1>
<p>For the purpose of homogeneization between plots, for the different phylogenetic trees, palettes and grouping helpers are defined in separate files which are sourced in this script.</p>
<ul>
<li>palettes are defined in the file <code>AI_palettes.R</code></li>
</ul>
</section>
<section id="importing-phylogenetic-tree-preparing-the-tree-and-metdata-for-plotting" class="level1">
<h1>Importing phylogenetic tree, preparing the tree and metdata for plotting</h1>
<section id="rooting-the-tree" class="level2">
<h2 class="anchored" data-anchor-id="rooting-the-tree">Rooting the tree</h2>
<ul>
<li><p>importing the tree and cleaning the tip labels (removing extension .fasta)</p></li>
<li><p>NB we used the consensus tree</p></li>
</ul>
<ul>
<li>fast plotting: we see that 2 very long branches, this are the outgroup sequences that we will use for rooting. We root those and remove them from the tree.</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N5_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>rooting the tree, and dropping the two tips that correspond to the long branch</li>
</ul>
<p>Verification:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N5_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="no-duplicate-sequences-to-integrate-in-this-tree" class="level2">
<h2 class="anchored" data-anchor-id="no-duplicate-sequences-to-integrate-in-this-tree">No duplicate sequences to integrate in this tree</h2>
<ul>
<li>creating table with metadata</li>
</ul>
<ul>
<li>recheck that we only have the correct subtypes included</li>
</ul>
<div class="cell">
<pre><code>#&gt; # A tibble: 2 × 2
#&gt;   subtype     n
#&gt;   &lt;chr&gt;   &lt;int&gt;
#&gt; 1 H5N1        1
#&gt; 2 H5N5       84</code></pre>
</div>
<p>This is ok that we have one H5N1 as its the reference/root</p>
<p>Adding filtering columns. We want to show where Norwegian sequences are,</p>
<p>For this purpose, we add columns that would allow to tag labels representing isolates from Norway, and those that contain White-tailed Eagles.</p>
<blockquote class="blockquote">
<p>Note here could have been simpler, but do as the same script as possible as H5N1_HA</p>
</blockquote>
</section>
<section id="exploration-of-the-phylogenetic-tree" class="level2">
<h2 class="anchored" data-anchor-id="exploration-of-the-phylogenetic-tree">Exploration of the phylogenetic tree</h2>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N5_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<!-- we do not collapse nodes for this tree -->
<div class="cell">
<pre><code>#&gt; Creating bootstrap field and names for nodes
#&gt; creating a simple plot
#&gt; merging bootstrap to plot data
#&gt; results are a list of a simple_plot and a tree with labelled nodes
#&gt; filtering the tree plot data to get only nodes
#&gt; checking if groups are monophyletic - without bootstrap support
#&gt; checking if groups are monophyletic - with bootstrap support
#&gt; adding edge data for plotting
#&gt; updating tree plot data and output
#&gt; Joining with `by = join_by(parent, node, branch.length, label, isTip, x, y,
#&gt; branch, angle, labels_bootstrap, node_depth)`
#&gt; Joining with `by = join_by(parent, node)`
#&gt; Returning the plot with the updated metadata. The tree with added node labels
#&gt; is in .$trees_nodeslab, and the simple plot is in .$simple_plot</code></pre>
</div>
<ul>
<li>exploring the structure</li>
</ul>
<ul>
<li>saving in individual objects (copy). We will reuse the plot data We need to add the tips in descendants from nodes - so its homogeneous for next function</li>
</ul>
<!-- data check

::: {.cell}

```
#> # A tibble: 83 × 15
#> # Rowwise: 
#>    parent  node branch.length label isTip        x     y   branch angle
#>     <int> <int>         <dbl> <chr> <lgl>    <dbl> <dbl>    <dbl> <dbl>
#>  1     85    85    0          N1    FALSE 0         12.2 0         52.4
#>  2     85    86    0.000682   N2    FALSE 0.000682  23.4 0.000341 100. 
#>  3     86    87    0.00000214 N3    FALSE 0.000684  39.7 0.000683 170. 
#>  4     87    88    0.00000214 N4    FALSE 0.000686  42.9 0.000685 184. 
#>  5     88    89    0.00118    N5    FALSE 0.00187   47.9 0.00128  205. 
#>  6     89    90    0.00178    N6    FALSE 0.00364   54.0 0.00275  232. 
#>  7     90    91    0.000990   N7    FALSE 0.00463   57.2 0.00414  245. 
#>  8     91    92    0.000489   N8    FALSE 0.00512   60.3 0.00488  259. 
#>  9     92    93    0.00148    N9    FALSE 0.00660   65.6 0.00586  281. 
#> 10     93    94    0.000587   N10   FALSE 0.00718   73.1 0.00689  313. 
#>    labels_bootstrap node_depth descendants_tips_from_node isMono_noboot isMono
#>    <chr>                 <dbl> <list>                     <lgl>         <lgl> 
#>  1 ""                       84 <chr [84]>                 TRUE          NA    
#>  2 "82"                     83 <chr [83]>                 TRUE          FALSE 
#>  3 "19"                     49 <chr [49]>                 TRUE          FALSE 
#>  4 "38"                     47 <chr [47]>                 TRUE          FALSE 
#>  5 "94"                     46 <chr [46]>                 TRUE          FALSE 
#>  6 "96"                     35 <chr [35]>                 TRUE          TRUE  
#>  7 "87"                     31 <chr [31]>                 TRUE          FALSE 
#>  8 "87"                     30 <chr [30]>                 TRUE          FALSE 
#>  9 "85"                     29 <chr [29]>                 TRUE          FALSE 
#> 10 "82"                     15 <chr [15]>                 TRUE          FALSE 
#>    edge_num
#>       <int>
#>  1       NA
#>  2        1
#>  3        2
#>  4        3
#>  5        4
#>  6        5
#>  7        6
#>  8        7
#>  9        8
#> 10        9
#> # ℹ 73 more rows
```
:::


Checking data - ok it has nested the descendants from nodes 

::: {.cell}

```
#> # A tibble: 84 × 1
#>    descendants_tips_from_node
#>    <chr>                     
#>  1 EPI_ISL_1123358           
#>  2 EPI_ISL_644737            
#>  3 EPI_ISL_12028892          
#>  4 2022_04_13257_1           
#>  5 2022_07_1140_1            
#>  6 2022_07_1141_1            
#>  7 EPI_ISL_18455200          
#>  8 2022_07_1233_1            
#>  9 2022_07_446               
#> 10 2022_07_693_1             
#> # ℹ 74 more rows
```
:::

-->
<section id="adding-metadata-to-nodes-and-tips" class="level3">
<h3 class="anchored" data-anchor-id="adding-metadata-to-nodes-and-tips">2. Adding metadata to nodes and tips</h3>
<blockquote class="blockquote">
<p>for consistency with previous script - we also add tipmetalink</p>
</blockquote>
<ul>
<li>linking metadata to the tree data for the selected column by creating nested dataframes for each variable and creating contingency tables for each nested variable</li>
</ul>
<p>We create contingency tables and nested variables that represent the metadata for each node and tip (label column).</p>
<p>Note here, we come back to the original metadata table, because we have the list of tips that is represented at each label (tip or node)</p>
<p>Here we do not want to use interaction between factors</p>
<p>Here is how it looks like</p>
<div class="cell">
<pre><code>#&gt;  [1] "parent"                     "node"                      
#&gt;  [3] "branch.length"              "label"                     
#&gt;  [5] "isTip"                      "x"                         
#&gt;  [7] "y"                          "branch"                    
#&gt;  [9] "angle"                      "labels_bootstrap"          
#&gt; [11] "node_depth"                 "descendants_tips_from_node"
#&gt; [13] "isMono_noboot"              "isMono"                    
#&gt; [15] "edge_num"                   "tip_metalink"              
#&gt; [17] "nested_country"             "ctg_country"               
#&gt; [19] "nested_Host"                "ctg_Host"                  
#&gt; [21] "nested_Host_latin"          "ctg_Host_latin"            
#&gt; [23] "nested_year"                "ctg_year"                  
#&gt; [25] "nested_continent"           "ctg_continent"             
#&gt; [27] "nested_Host_family"         "ctg_Host_family"</code></pre>
</div>
<div class="cell">
<pre><code>#&gt; # A tibble: 167 × 28
#&gt; # Rowwise: 
#&gt;    parent  node branch.length label            isTip       x     y  branch angle
#&gt;     &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;            &lt;lgl&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
#&gt;  1    101     1    0          EPI_ISL_18455200 TRUE  0.00837    83 0.00837  356.
#&gt;  2    101     2    0          2022_07_1233_1   TRUE  0.00837    84 0.00837  360 
#&gt;  3    100     3    0.00000214 2022_07_1141_1   TRUE  0.00837    82 0.00837  351.
#&gt;  4    102     4    0.00000214 2022_07_446      TRUE  0.00837    80 0.00837  343.
#&gt;  5    102     5    0.00000214 2022_07_693_1    TRUE  0.00837    81 0.00837  347.
#&gt;  6     98     6    0.000587   2022_07_1140_1   TRUE  0.00895    79 0.00865  339.
#&gt;  7    103     7    0.00000214 2022_07_448      TRUE  0.00836    77 0.00836  330 
#&gt;  8    103     8    0.00000214 2022_07_981_1    TRUE  0.00836    78 0.00836  334.
#&gt;  9    104     9    0.00000214 2022_07_1137_1   TRUE  0.00778    75 0.00778  321.
#&gt; 10    104    10    0.00000214 2022_07_895_1    TRUE  0.00778    76 0.00778  326.
#&gt;    labels_bootstrap node_depth descendants_tips_from_node isMono_noboot isMono
#&gt;    &lt;chr&gt;                 &lt;dbl&gt; &lt;list&gt;                     &lt;lgl&gt;         &lt;lgl&gt; 
#&gt;  1 EPI_ISL_18455200          1 &lt;chr [1]&gt;                  NA            NA    
#&gt;  2 2022_07_1233_1            1 &lt;chr [1]&gt;                  NA            NA    
#&gt;  3 2022_07_1141_1            1 &lt;chr [1]&gt;                  NA            NA    
#&gt;  4 2022_07_446               1 &lt;chr [1]&gt;                  NA            NA    
#&gt;  5 2022_07_693_1             1 &lt;chr [1]&gt;                  NA            NA    
#&gt;  6 2022_07_1140_1            1 &lt;chr [1]&gt;                  NA            NA    
#&gt;  7 2022_07_448               1 &lt;chr [1]&gt;                  NA            NA    
#&gt;  8 2022_07_981_1             1 &lt;chr [1]&gt;                  NA            NA    
#&gt;  9 2022_07_1137_1            1 &lt;chr [1]&gt;                  NA            NA    
#&gt; 10 2022_07_895_1             1 &lt;chr [1]&gt;                  NA            NA    
#&gt;    edge_num tip_metalink nested_country   ctg_country      nested_Host     
#&gt;       &lt;int&gt; &lt;list&gt;       &lt;list&gt;           &lt;list&gt;           &lt;list&gt;          
#&gt;  1       17 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  2       18 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  3       19 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  4       21 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  5       22 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  6       23 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  7       25 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  8       26 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  9       28 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt; 10       29 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;    ctg_Host         nested_Host_latin ctg_Host_latin   nested_year     
#&gt;    &lt;list&gt;           &lt;list&gt;            &lt;list&gt;           &lt;list&gt;          
#&gt;  1 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  2 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  3 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  4 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  5 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  6 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  7 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  8 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  9 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt; 10 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;    ctg_year         nested_continent ctg_continent    nested_Host_family
#&gt;    &lt;list&gt;           &lt;list&gt;           &lt;list&gt;           &lt;list&gt;            
#&gt;  1 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  2 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  3 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  4 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  5 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  6 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  7 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  8 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  9 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt; 10 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;    ctg_Host_family 
#&gt;    &lt;list&gt;          
#&gt;  1 &lt;tibble [1 × 4]&gt;
#&gt;  2 &lt;tibble [1 × 4]&gt;
#&gt;  3 &lt;tibble [1 × 4]&gt;
#&gt;  4 &lt;tibble [1 × 4]&gt;
#&gt;  5 &lt;tibble [1 × 4]&gt;
#&gt;  6 &lt;tibble [1 × 4]&gt;
#&gt;  7 &lt;tibble [1 × 4]&gt;
#&gt;  8 &lt;tibble [1 × 4]&gt;
#&gt;  9 &lt;tibble [1 × 4]&gt;
#&gt; 10 &lt;tibble [1 × 4]&gt;
#&gt; # ℹ 157 more rows</code></pre>
</div>
<p>Selecting one column where several tips are respresented and showing an example of how the nested variable looks like, using eg. nested country variable</p>
<blockquote class="blockquote">
<p>note for this data set its only nodes</p>
</blockquote>
<div class="cell">
<pre><code>#&gt; # A tibble: 84 × 3
#&gt;    label            country          pos
#&gt;    &lt;chr&gt;            &lt;chr&gt;          &lt;int&gt;
#&gt;  1 EPI_ISL_1123358  United Kingdom     1
#&gt;  2 EPI_ISL_644737   Denmark            2
#&gt;  3 EPI_ISL_12028892 Norway             3
#&gt;  4 2022_04_13257_1  Norway             4
#&gt;  5 2022_07_1140_1   Norway             5
#&gt;  6 2022_07_1141_1   Norway             6
#&gt;  7 EPI_ISL_18455200 Norway             7
#&gt;  8 2022_07_1233_1   Norway             8
#&gt;  9 2022_07_446      Norway             9
#&gt; 10 2022_07_693_1    Norway            10
#&gt; # ℹ 74 more rows</code></pre>
</div>
<p><code>pos</code> is the index, in case we need to refer to it. <code>label</code> is the label of the sequences that are represented</p>
<p>Now we can look at the contingency table for the same variable</p>
<div class="cell">
<pre><code>#&gt; # A tibble: 17 × 4
#&gt;    country            `freq_levels@country` `nb_levels@country`   pos
#&gt;    &lt;chr&gt;                              &lt;int&gt;               &lt;int&gt; &lt;int&gt;
#&gt;  1 United Kingdom                         3                  17     1
#&gt;  2 Denmark                                1                  17     2
#&gt;  3 Norway                                31                  17     3
#&gt;  4 Russian Federation                     9                  17     4
#&gt;  5 Romania                                8                  17     5
#&gt;  6 Bulgaria                               2                  17     6
#&gt;  7 Poland                                 1                  17     7
#&gt;  8 Italy                                  1                  17     8
#&gt;  9 Germany                                2                  17     9
#&gt; 10 Slovenia                               1                  17    10
#&gt; 11 Sweden                                17                  17    11
#&gt; 12 Netherlands                            1                  17    12
#&gt; 13 Belgium                                1                  17    13
#&gt; 14 Slovakia                               3                  17    14
#&gt; 15 Hungary                                1                  17    15
#&gt; 16 Austria                                1                  17    16
#&gt; 17 Czech Republic                         1                  17    17</code></pre>
</div>
<p><code>freq_levels@country</code> : is the number of observations in each country <code>nb_levels@country</code> : is common in the nested table, its the number of levels of the variable country</p>
<ul>
<li>NOW: we can reinsert the augmented data to the plot</li>
</ul>
</section>
<section id="filtering-based-on-criteria-based-on-information-from-nested-metadata" class="level3">
<h3 class="anchored" data-anchor-id="filtering-based-on-criteria-based-on-information-from-nested-metadata">3. Filtering based on criteria based on information from nested metadata</h3>
<ul>
<li>Filtering based on criteria that are found in the nested metadata that is attached to the tree plot_data</li>
</ul>
<p>We add a tag - if there is at least one isolate that is represented by a label in the phylogenetic tree has been collected in Norway</p>
<div class="cell">
<pre><code>#&gt;  [1] "parent"                     "node"                      
#&gt;  [3] "branch.length"              "label"                     
#&gt;  [5] "isTip"                      "x"                         
#&gt;  [7] "y"                          "branch"                    
#&gt;  [9] "angle"                      "labels_bootstrap"          
#&gt; [11] "node_depth"                 "descendants_tips_from_node"
#&gt; [13] "isMono_noboot"              "isMono"                    
#&gt; [15] "edge_num"                   "tip_metalink"              
#&gt; [17] "nested_country"             "ctg_country"               
#&gt; [19] "nested_Host"                "ctg_Host"                  
#&gt; [21] "nested_Host_latin"          "ctg_Host_latin"            
#&gt; [23] "nested_year"                "ctg_year"                  
#&gt; [25] "nested_continent"           "ctg_continent"             
#&gt; [27] "nested_Host_family"         "ctg_Host_family"           
#&gt; [29] "inNorway"                   "isWTE"                     
#&gt; [31] "category_highlight"</code></pre>
</div>
</section>
<section id="test-visualisation" class="level3">
<h3 class="anchored" data-anchor-id="test-visualisation">4. Test visualisation</h3>
<p>We can visualize which nodes can be collapsed and at which depths We can try to vary the depth of notes to be collapsed</p>
<div class="cell">
<pre><code>#&gt;  [1] "parent"                     "node"                      
#&gt;  [3] "branch.length"              "label"                     
#&gt;  [5] "isTip"                      "x"                         
#&gt;  [7] "y"                          "branch"                    
#&gt;  [9] "angle"                      "labels_bootstrap"          
#&gt; [11] "node_depth"                 "descendants_tips_from_node"
#&gt; [13] "isMono_noboot"              "isMono"                    
#&gt; [15] "edge_num"                   "tip_metalink"              
#&gt; [17] "nested_country"             "ctg_country"               
#&gt; [19] "nested_Host"                "ctg_Host"                  
#&gt; [21] "nested_Host_latin"          "ctg_Host_latin"            
#&gt; [23] "nested_year"                "ctg_year"                  
#&gt; [25] "nested_continent"           "ctg_continent"             
#&gt; [27] "nested_Host_family"         "ctg_Host_family"           
#&gt; [29] "inNorway"                   "isWTE"                     
#&gt; [31] "category_highlight"</code></pre>
</div>
<p>We add the node labels so we can know which ones we can collapse afterwards</p>
<div class="cell">
<pre><code>#&gt; Warning: There were 84 warnings in `filter()`.
#&gt; The first warning was:
#&gt; ℹ In argument: `as.numeric(labels_bootstrap) &gt;= 95`.
#&gt; ℹ In row 1.
#&gt; Caused by warning:
#&gt; ! NAs introduced by coercion
#&gt; ℹ Run `dplyr::last_dplyr_warnings()` to see the 83 remaining warnings.</code></pre>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N5_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>ok here we see clearily that there is no large group to collapse to make it worsewhile</p>
</blockquote>
<div class="cell">
<pre><code>#&gt; Warning: There were 84 warnings in `filter()`.
#&gt; The first warning was:
#&gt; ℹ In argument: `as.numeric(labels_bootstrap) &gt;= 95`.
#&gt; ℹ In row 1.
#&gt; Caused by warning:
#&gt; ! NAs introduced by coercion
#&gt; ℹ Run `dplyr::last_dplyr_warnings()` to see the 83 remaining warnings.</code></pre>
</div>
</section>
</section>
</section>
<section id="components-for-final-visualisation" class="level1">
<h1>Components for final visualisation</h1>
<p>We will create the global circular tree And the two groups where White-tailed eagle are found to create trees that can be looked at more in detail.</p>
</section>
<section id="circular-tree" class="level1">
<h1>Circular tree</h1>
<p>We restart from the rooted tree with node labels</p>
<p>Circular metadata will be the at the last plots, except we do not add the coordinates</p>
<p>We had some plotting parameters to be able to view better</p>
<ul>
<li>Adding annotation around the tree <!-- here they wanted by country not by group of country --></li>
</ul>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N5_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><em>Figure caption:</em> H5N5 HA segment. ML phylogenetic tree. The tree is rooted, but the outgroup is not shown, due to large distance to the body of the tree.</p>
<section id="rectangular-plot" class="level2">
<h2 class="anchored" data-anchor-id="rectangular-plot">Rectangular plot</h2>
<ul>
<li>preparing the data for the pannels views</li>
</ul>
<ul>
<li>ploting the tree</li>
</ul>
<div class="cell">
<pre><code>#&gt; The min value of positional variable is 0.0004974486 and the max value is 0.0106837522.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 1 sequences positions for your unnested variable
#&gt; The min value of positional variable is 0.0004974486 and the max value is 0.0106837522.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 1 sequences positions for your unnested variable
#&gt; The min value of positional variable is 0.0004974486 and the max value is 0.0106837522.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 1 sequences positions for your unnested variable
#&gt; The min value of positional variable is 0.0004974486 and the max value is 0.0106837522.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 1 sequences positions for your unnested variable</code></pre>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N5_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Now we need to adjust the size of the panels</li>
</ul>
<div class="cell">
<pre><code>#&gt; The min value of positional variable is 0.0004974486 and the max value is 0.0106837522.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 1 sequences positions for your unnested variable</code></pre>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N5_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><em>Figure Caption:</em> H5N5 HA segment. ML phylogenetic tree. The tree is rooted but the root is not shown as it is too distant from the rest of the tree.</p>
</section>
</section>
<section id="saving-environment" class="level1">
<h1>Saving environment</h1>
<p>To continue analyses that are not finished</p>
<p>Metadata for all H5N5 HA segment isolates included</p>
<p>This dataset is representative for 85 isolates. This includes 85 isolates with unique sequences, including 1 sequence as outgroup.</p>
<p>Some isolates have been submited to GISAID since we started this analysis, therefore the complement of GISAID_ID will be completed manually.</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>