<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Eve Zeyl Fiskebeck">
<meta name="dcterms.date" content="2024-02-05">

<title>SEP H5N1 HA Phylogenetic tree</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="SEP_H5N1_HA_phylogenetic_tree_files/libs/clipboard/clipboard.min.js"></script>
<script src="SEP_H5N1_HA_phylogenetic_tree_files/libs/quarto-html/quarto.js"></script>
<script src="SEP_H5N1_HA_phylogenetic_tree_files/libs/quarto-html/popper.min.js"></script>
<script src="SEP_H5N1_HA_phylogenetic_tree_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="SEP_H5N1_HA_phylogenetic_tree_files/libs/quarto-html/anchor.min.js"></script>
<link href="SEP_H5N1_HA_phylogenetic_tree_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="SEP_H5N1_HA_phylogenetic_tree_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="SEP_H5N1_HA_phylogenetic_tree_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="SEP_H5N1_HA_phylogenetic_tree_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="SEP_H5N1_HA_phylogenetic_tree_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">SEP H5N1 HA Phylogenetic tree</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Eve Zeyl Fiskebeck </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 5, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="setup" class="level1">
<h1>Setup</h1>
<ul>
<li>sourcing scripts and functions</li>
</ul>
<p>Note: if you are modifying the script and want to continue to work after, you can load a previous environment (don’t forget to save it at the end of your work).</p>
<p>Loads the environment (to continue previous work)</p>
</section>
<section id="preparation-of-metadata-specific-to-the-tree" class="level1">
<h1>Preparation of metadata specific to the tree</h1>
<p>This is the metadata that we combined to do all the trees, and where the host family has been added manually.</p>
<div class="cell">
<pre><code>#&gt; Rows: 5,129
#&gt; Columns: 28
#&gt; $ label                &lt;chr&gt; "EPI_ISL_1224989", "EPI_ISL_7267244", "EPI_ISL_11…
#&gt; $ Isolate_Id           &lt;chr&gt; "EPI_ISL_1224989", "EPI_ISL_7267244", "EPI_ISL_11…
#&gt; $ subtype              &lt;chr&gt; "H5N1", "H5N1", "H5N1", "H5N1", "H5N1", "H5N1", "…
#&gt; $ continent            &lt;chr&gt; "Europe", "Europe", "Europe", "Europe", "Europe",…
#&gt; $ country              &lt;chr&gt; "Netherlands", "Netherlands", "Denmark", "United …
#&gt; $ Host                 &lt;chr&gt; "Buteo buteo", "Buteo buteo", "Buteo buteo", "But…
#&gt; $ year                 &lt;int&gt; 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2…
#&gt; $ month                &lt;int&gt; 2, 11, 10, 11, 11, 11, 12, 12, 1, 5, 11, 12, 12, …
#&gt; $ day                  &lt;int&gt; 1, 12, 25, 2, 20, 29, 15, 24, 9, 7, 25, 26, 27, 3…
#&gt; $ Locality             &lt;chr&gt; "NA", "NA", "NA", "NA", "NA", "NA", "NA", "NA", "…
#&gt; $ Locality_code        &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
#&gt; $ lon                  &lt;dbl&gt; 5.291266, 5.291266, 9.501785, -3.435973, -3.43597…
#&gt; $ lat                  &lt;dbl&gt; 52.13263, 52.13263, 56.26392, 55.37805, 55.37805,…
#&gt; $ Host_latin           &lt;chr&gt; "Buteo buteo", "Buteo buteo", "Buteo buteo", "But…
#&gt; $ precision_Host_latin &lt;chr&gt; "NA", "NA", "NA", "NA", "NA", "NA", "NA", "NA", "…
#&gt; $ comment              &lt;chr&gt; "NA", "NA", "NA", "NA", "NA", "NA", "NA", "NA", "…
#&gt; $ artkode              &lt;dbl&gt; 5080102009, 5080102009, 5080102009, 5080102009, 5…
#&gt; $ art                  &lt;chr&gt; "Musvåk", "Musvåk", "Musvåk", "Musvåk", "Musvåk",…
#&gt; $ efsakode             &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
#&gt; $ efsanavn             &lt;chr&gt; "Common Buzzard (as animal)", "Common Buzzard (as…
#&gt; $ norsk_navn           &lt;chr&gt; "Musvåk", "Musvåk", "Musvåk", "Musvåk", "Musvåk",…
#&gt; $ engelsk_navn         &lt;chr&gt; "Common Buzzard", "Common Buzzard", "Common Buzza…
#&gt; $ slekt                &lt;chr&gt; "Buteo", "Buteo", "Buteo", "Buteo", "Buteo", "But…
#&gt; $ familie              &lt;chr&gt; "Accipitridae", "Accipitridae", "Accipitridae", "…
#&gt; $ orden                &lt;chr&gt; "ACCIPITRIFORMES", "ACCIPITRIFORMES", "ACCIPITRIF…
#&gt; $ latinsk_navn         &lt;chr&gt; "NA", "NA", "NA", "NA", "NA", "NA", "NA", "NA", "…
#&gt; $ Host_family          &lt;chr&gt; "Accipitridae", "Accipitridae", "Accipitridae", "…
#&gt; $ Isolate_Name         &lt;chr&gt; "A/common buzzard/Netherlands/21022834-002/2021",…</code></pre>
</div>
<p>We select only the columns that we will use, to reduce complexity</p>
<ul>
<li>List of duplicate sequences that were removed prior building the phylogenetic tree for strain H5N1, segment HA</li>
</ul>
</section>
<section id="information" class="level1">
<h1>Information:</h1>
<p>For the purpose of homogeneization between plots, for the different phylogenetic trees, palettes and grouping helpers are defined in separate files which are sourced in this script.</p>
<ul>
<li>palettes are defined in the file <code>AI_palettes.R</code></li>
</ul>
<!-- reasoning detail and putting that in place in 2024-01-02_SEP_HA_final_tree_script.Rmd
last commit 2024-01-30 in the pilot project. After started to concise in functions.
-->
</section>
<section id="importing-phylogenetic-tree-preparing-the-tree-and-metdata-for-plotting" class="level1">
<h1>Importing phylogenetic tree, preparing the tree and metdata for plotting</h1>
<section id="rooting-the-tree" class="level2">
<h2 class="anchored" data-anchor-id="rooting-the-tree">Rooting the tree</h2>
<ul>
<li><p>importing the tree and cleaning the tip labels (removing extension .fasta)</p></li>
<li><p>NB we used the consensus tree</p></li>
</ul>
<ul>
<li>fast plotting: we see that 2 very long branches, this are the outgroup sequences that we will use for rooting. We root those and remove them from the tree.</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>rooting the tree, and dropping the two tips that correspond to the long branch</li>
</ul>
<p>Verification:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="integrating-information-from-duplicate-sequences" class="level2">
<h2 class="anchored" data-anchor-id="integrating-information-from-duplicate-sequences">Integrating information from duplicate sequences</h2>
<ul>
<li>because we had identical sequences, that were removed prior reconstructing the phylogenetic tree with IQTREE, we need to re-integrate the metadata and information that was contained in those sequences.</li>
</ul>
<p>Creating a dataframe from the list of duplicates produced by seqkit:</p>
<div class="cell">
<pre><code>#&gt; Rows: 293 Columns: 2
#&gt; ── Column specification ────────────────────────────────────────────────────────
#&gt; Delimiter: "\t"
#&gt; chr (1): duplicates
#&gt; dbl (1): ndub
#&gt; 
#&gt; ℹ Use `spec()` to retrieve the full column specification for this data.
#&gt; ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
#&gt; # A tibble: 6 × 5
#&gt;   duppos represented_seq_label dupgrp nb_represented_sequences label          
#&gt;    &lt;int&gt; &lt;chr&gt;                  &lt;int&gt;                    &lt;int&gt; &lt;chr&gt;          
#&gt; 1      1 EPI_ISL_5804698            1                       58 EPI_ISL_5804698
#&gt; 2      2 EPI_ISL_6025676            1                       58 EPI_ISL_5804698
#&gt; 3      3 EPI_ISL_5945412            1                       58 EPI_ISL_5804698
#&gt; 4      4 EPI_ISL_6101847            1                       58 EPI_ISL_5804698
#&gt; 5      5 EPI_ISL_6101868            1                       58 EPI_ISL_5804698
#&gt; 6      6 EPI_ISL_6761001            1                       58 EPI_ISL_5804698</code></pre>
</div>
<p>Matching duplicate information for plotting will require a tree plot and then merging. But first we create the metadata for the tree, that includes the duplicates metadata.</p>
<blockquote class="blockquote">
<p>We can do that for the general tree (as afterwards the information that is not in the tree is by labels)</p>
</blockquote>
<div class="cell">
<pre><code>#&gt; Joining with `by = join_by(label)`</code></pre>
</div>
<ul>
<li>recheck that we only have the correct subtypes included</li>
</ul>
<div class="cell">
<pre><code>#&gt; Joining with `by = join_by(label)`
#&gt; # A tibble: 1 × 2
#&gt;   subtype     n
#&gt;   &lt;chr&gt;   &lt;int&gt;
#&gt; 1 H5N1     2378</code></pre>
</div>
<p>Now we need to get the contingency tables for each label (which can represent one or several isolates when it is identical to other sequences</p>
<p>We obtain a nested table that can be matched to the labels of the tips of the tree</p>
<div class="cell">
<pre><code>#&gt; # A tibble: 6 × 12
#&gt; # Rowwise: 
#&gt;   label    duppos represented_seq_label dupgrp nb_represented_seque…¹ Isolate_Id
#&gt;   &lt;chr&gt;    &lt;list&gt; &lt;list&gt;                 &lt;dbl&gt;                  &lt;dbl&gt; &lt;chr&gt;     
#&gt; 1 EPI_ISL… &lt;dbl&gt;  &lt;chr [58]&gt;                 1                     58 EPI_ISL_5…
#&gt; 2 EPI_ISL… &lt;dbl&gt;  &lt;chr [58]&gt;                 1                     58 EPI_ISL_6…
#&gt; 3 EPI_ISL… &lt;dbl&gt;  &lt;chr [58]&gt;                 1                     58 EPI_ISL_5…
#&gt; 4 EPI_ISL… &lt;dbl&gt;  &lt;chr [58]&gt;                 1                     58 EPI_ISL_6…
#&gt; 5 EPI_ISL… &lt;dbl&gt;  &lt;chr [58]&gt;                 1                     58 EPI_ISL_6…
#&gt; 6 EPI_ISL… &lt;dbl&gt;  &lt;chr [58]&gt;                 1                     58 EPI_ISL_6…
#&gt; # ℹ abbreviated name: ¹​nb_represented_sequences
#&gt; # ℹ 6 more variables: continent &lt;list&gt;, country &lt;list&gt;, year &lt;list&gt;,
#&gt; #   Host &lt;list&gt;, Host_latin &lt;list&gt;, Host_family &lt;list&gt;</code></pre>
</div>
<p>Adding filtering columns, using the nested data. We want to show where Norwegian sequences are,</p>
<p>but they can be duplicated sequences and thus not the primary labeled sequences on the tree.</p>
<p>For this purpose, we add columns that would allow to tag labels representing isolates from Norway, and those that contain White-tailed Eagles.</p>
</section>
<section id="exploration-of-the-phylogenetic-tree" class="level2">
<h2 class="anchored" data-anchor-id="exploration-of-the-phylogenetic-tree">Exploration of the phylogenetic tree</h2>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="collapsing-nodes" class="level1">
<h1>Collapsing nodes</h1>
<section id="a.-preparation-to-collapse-nodes" class="level2">
<h2 class="anchored" data-anchor-id="a.-preparation-to-collapse-nodes">A. Preparation to collapse nodes</h2>
<p>We start from scratch.</p>
<section id="creating-the-plot-and-tree-data-adding-information-to-the-nodes-monophyly-depth-nodes-all-descendants" class="level3">
<h3 class="anchored" data-anchor-id="creating-the-plot-and-tree-data-adding-information-to-the-nodes-monophyly-depth-nodes-all-descendants">1. creating the plot and tree data: adding information to the nodes (monophyly, depth nodes, all descendants)</h3>
<ul>
<li><p>tagging monophyletic nodes with bootstrap support (&gt;=95), adding node labels to be able to retrieve information, moving bootstrap values to another column, and computing all nodes descendants and node depth (this can take some time)</p></li>
<li><p>this create the simple plot at the same time</p></li>
</ul>
<div class="cell">
<pre><code>#&gt; Creating bootstrap field and names for nodes
#&gt; creating a simple plot
#&gt; merging bootstrap to plot data
#&gt; results are a list of a simple_plot and a tree with labelled nodes
#&gt; filtering the tree plot data to get only nodes
#&gt; checking if groups are monophyletic - without bootstrap support
#&gt; checking if groups are monophyletic - with bootstrap support
#&gt; adding edge data for plotting
#&gt; updating tree plot data and output
#&gt; Joining with `by = join_by(parent, node, branch.length, label, isTip, x, y,
#&gt; branch, angle, labels_bootstrap, node_depth)`
#&gt; Joining with `by = join_by(parent, node)`
#&gt; Returning the plot with the updated metadata. The tree with added node labels
#&gt; is in .$trees_nodeslab, and the simple plot is in .$simple_plot</code></pre>
</div>
<ul>
<li>exploring the structure</li>
</ul>
<ul>
<li>saving in individual objects (copy). We will reuse the plot data</li>
</ul>
</section>
<section id="adding-metadata-to-nodes-and-tips---integration-of-duplicates-metadata" class="level3">
<h3 class="anchored" data-anchor-id="adding-metadata-to-nodes-and-tips---integration-of-duplicates-metadata">2. Adding metadata to nodes and tips - integration of duplicates metadata</h3>
<ul>
<li>we add a column where all the isolates (duplicated and non duplicated) that are represented at a tip and node are listed.</li>
</ul>
<p>We call this column tip_metalink This column will be used to make the join between the tree and the metadata that is represented at each tip or node. It uses the metadata table for duplicates.</p>
<p>(This can take some time). &gt; Opps check defaults for the function if you renamed the objects!</p>
<p>Note that isMono_x has no information for the tips (this is normal).</p>
<div class="cell">
<pre><code>#&gt; # A tibble: 2,825 × 16
#&gt; # Rowwise: 
#&gt;    parent  node branch.length label            isTip      x     y branch angle
#&gt;     &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;            &lt;lgl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
#&gt;  1   1471     1    0.000595   EPI_ISL_18455202 TRUE  0.0119   494 0.0116  125.
#&gt;  2   1471     2    0.000595   EPI_ISL_14573031 TRUE  0.0119   495 0.0116  126.
#&gt;  3   1470     3    0.00000253 2022_07_1236_1   TRUE  0.0113   493 0.0113  125.
#&gt;  4   1469     4    0.00000253 EPI_ISL_15267026 TRUE  0.0113   492 0.0113  125.
#&gt;  5   1468     5    0.000596   EPI_ISL_14811929 TRUE  0.0119   491 0.0116  125.
#&gt;  6   1467     6    0.000595   2022_07_1143_1   TRUE  0.0119   490 0.0116  124.
#&gt;  7   1466     7    0.00000253 2022_04_20940_1  TRUE  0.0107   489 0.0107  124.
#&gt;  8   1465     8    0.00119    EPI_ISL_15267024 TRUE  0.0119   488 0.0113  124.
#&gt;  9   1473     9    0.00119    EPI_ISL_16507108 TRUE  0.0125   486 0.0119  123.
#&gt; 10   1473    10    0.00239    EPI_ISL_15579544 TRUE  0.0136   487 0.0125  124.
#&gt;    labels_bootstrap node_depth descendants_tips_from_node isMono_noboot isMono
#&gt;    &lt;chr&gt;                 &lt;dbl&gt; &lt;list&gt;                     &lt;lgl&gt;         &lt;lgl&gt; 
#&gt;  1 EPI_ISL_18455202          1 &lt;NULL&gt;                     NA            NA    
#&gt;  2 EPI_ISL_14573031          1 &lt;NULL&gt;                     NA            NA    
#&gt;  3 2022_07_1236_1            1 &lt;NULL&gt;                     NA            NA    
#&gt;  4 EPI_ISL_15267026          1 &lt;NULL&gt;                     NA            NA    
#&gt;  5 EPI_ISL_14811929          1 &lt;NULL&gt;                     NA            NA    
#&gt;  6 2022_07_1143_1            1 &lt;NULL&gt;                     NA            NA    
#&gt;  7 2022_04_20940_1           1 &lt;NULL&gt;                     NA            NA    
#&gt;  8 EPI_ISL_15267024          1 &lt;NULL&gt;                     NA            NA    
#&gt;  9 EPI_ISL_16507108          1 &lt;NULL&gt;                     NA            NA    
#&gt; 10 EPI_ISL_15579544          1 &lt;NULL&gt;                     NA            NA    
#&gt;    edge_num tip_metalink
#&gt;       &lt;int&gt; &lt;list&gt;      
#&gt;  1       52 &lt;chr [2]&gt;   
#&gt;  2       53 &lt;chr [1]&gt;   
#&gt;  3       54 &lt;chr [1]&gt;   
#&gt;  4       55 &lt;chr [2]&gt;   
#&gt;  5       56 &lt;chr [1]&gt;   
#&gt;  6       57 &lt;chr [1]&gt;   
#&gt;  7       58 &lt;chr [1]&gt;   
#&gt;  8       59 &lt;chr [1]&gt;   
#&gt;  9       62 &lt;chr [1]&gt;   
#&gt; 10       63 &lt;chr [1]&gt;   
#&gt; # ℹ 2,815 more rows</code></pre>
</div>
<ul>
<li>linking metadata to the tree data for the selected column by creating nested dataframes for each variable and creating contingency tables for each nested variable</li>
</ul>
<p>We create contingency tables and nested variables that represent the metadata for each node and tip (label column).</p>
<p>Note here, we come back to the original metadata table, because we have the list of tips that is represented at each label (tip or node)</p>
<p>Here we do not want to use interaction between factors</p>
<p>Here is how it looks like</p>
<div class="cell">
<pre><code>#&gt;  [1] "parent"                     "node"                      
#&gt;  [3] "branch.length"              "label"                     
#&gt;  [5] "isTip"                      "x"                         
#&gt;  [7] "y"                          "branch"                    
#&gt;  [9] "angle"                      "labels_bootstrap"          
#&gt; [11] "node_depth"                 "descendants_tips_from_node"
#&gt; [13] "isMono_noboot"              "isMono"                    
#&gt; [15] "edge_num"                   "tip_metalink"              
#&gt; [17] "nested_country"             "ctg_country"               
#&gt; [19] "nested_Host"                "ctg_Host"                  
#&gt; [21] "nested_Host_latin"          "ctg_Host_latin"            
#&gt; [23] "nested_year"                "ctg_year"                  
#&gt; [25] "nested_continent"           "ctg_continent"             
#&gt; [27] "nested_Host_family"         "ctg_Host_family"</code></pre>
</div>
<div class="cell">
<pre><code>#&gt; # A tibble: 2,825 × 28
#&gt; # Rowwise: 
#&gt;    parent  node branch.length label            isTip      x     y branch angle
#&gt;     &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;            &lt;lgl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
#&gt;  1   1471     1    0.000595   EPI_ISL_18455202 TRUE  0.0119   494 0.0116  125.
#&gt;  2   1471     2    0.000595   EPI_ISL_14573031 TRUE  0.0119   495 0.0116  126.
#&gt;  3   1470     3    0.00000253 2022_07_1236_1   TRUE  0.0113   493 0.0113  125.
#&gt;  4   1469     4    0.00000253 EPI_ISL_15267026 TRUE  0.0113   492 0.0113  125.
#&gt;  5   1468     5    0.000596   EPI_ISL_14811929 TRUE  0.0119   491 0.0116  125.
#&gt;  6   1467     6    0.000595   2022_07_1143_1   TRUE  0.0119   490 0.0116  124.
#&gt;  7   1466     7    0.00000253 2022_04_20940_1  TRUE  0.0107   489 0.0107  124.
#&gt;  8   1465     8    0.00119    EPI_ISL_15267024 TRUE  0.0119   488 0.0113  124.
#&gt;  9   1473     9    0.00119    EPI_ISL_16507108 TRUE  0.0125   486 0.0119  123.
#&gt; 10   1473    10    0.00239    EPI_ISL_15579544 TRUE  0.0136   487 0.0125  124.
#&gt;    labels_bootstrap node_depth descendants_tips_from_node isMono_noboot isMono
#&gt;    &lt;chr&gt;                 &lt;dbl&gt; &lt;list&gt;                     &lt;lgl&gt;         &lt;lgl&gt; 
#&gt;  1 EPI_ISL_18455202          1 &lt;NULL&gt;                     NA            NA    
#&gt;  2 EPI_ISL_14573031          1 &lt;NULL&gt;                     NA            NA    
#&gt;  3 2022_07_1236_1            1 &lt;NULL&gt;                     NA            NA    
#&gt;  4 EPI_ISL_15267026          1 &lt;NULL&gt;                     NA            NA    
#&gt;  5 EPI_ISL_14811929          1 &lt;NULL&gt;                     NA            NA    
#&gt;  6 2022_07_1143_1            1 &lt;NULL&gt;                     NA            NA    
#&gt;  7 2022_04_20940_1           1 &lt;NULL&gt;                     NA            NA    
#&gt;  8 EPI_ISL_15267024          1 &lt;NULL&gt;                     NA            NA    
#&gt;  9 EPI_ISL_16507108          1 &lt;NULL&gt;                     NA            NA    
#&gt; 10 EPI_ISL_15579544          1 &lt;NULL&gt;                     NA            NA    
#&gt;    edge_num tip_metalink nested_country   ctg_country      nested_Host     
#&gt;       &lt;int&gt; &lt;list&gt;       &lt;list&gt;           &lt;list&gt;           &lt;list&gt;          
#&gt;  1       52 &lt;chr [2]&gt;    &lt;tibble [2 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [2 × 3]&gt;
#&gt;  2       53 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  3       54 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  4       55 &lt;chr [2]&gt;    &lt;tibble [2 × 3]&gt; &lt;tibble [2 × 4]&gt; &lt;tibble [2 × 3]&gt;
#&gt;  5       56 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  6       57 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  7       58 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  8       59 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  9       62 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt; 10       63 &lt;chr [1]&gt;    &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;    ctg_Host         nested_Host_latin ctg_Host_latin   nested_year     
#&gt;    &lt;list&gt;           &lt;list&gt;            &lt;list&gt;           &lt;list&gt;          
#&gt;  1 &lt;tibble [2 × 4]&gt; &lt;tibble [2 × 3]&gt;  &lt;tibble [2 × 4]&gt; &lt;tibble [2 × 3]&gt;
#&gt;  2 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  3 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  4 &lt;tibble [2 × 4]&gt; &lt;tibble [2 × 3]&gt;  &lt;tibble [2 × 4]&gt; &lt;tibble [2 × 3]&gt;
#&gt;  5 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  6 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  7 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  8 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;  9 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt; 10 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;
#&gt;    ctg_year         nested_continent ctg_continent    nested_Host_family
#&gt;    &lt;list&gt;           &lt;list&gt;           &lt;list&gt;           &lt;list&gt;            
#&gt;  1 &lt;tibble [1 × 4]&gt; &lt;tibble [2 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [2 × 3]&gt;  
#&gt;  2 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  3 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  4 &lt;tibble [1 × 4]&gt; &lt;tibble [2 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [2 × 3]&gt;  
#&gt;  5 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  6 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  7 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  8 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;  9 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt; 10 &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt; &lt;tibble [1 × 4]&gt; &lt;tibble [1 × 3]&gt;  
#&gt;    ctg_Host_family 
#&gt;    &lt;list&gt;          
#&gt;  1 &lt;tibble [2 × 4]&gt;
#&gt;  2 &lt;tibble [1 × 4]&gt;
#&gt;  3 &lt;tibble [1 × 4]&gt;
#&gt;  4 &lt;tibble [2 × 4]&gt;
#&gt;  5 &lt;tibble [1 × 4]&gt;
#&gt;  6 &lt;tibble [1 × 4]&gt;
#&gt;  7 &lt;tibble [1 × 4]&gt;
#&gt;  8 &lt;tibble [1 × 4]&gt;
#&gt;  9 &lt;tibble [1 × 4]&gt;
#&gt; 10 &lt;tibble [1 × 4]&gt;
#&gt; # ℹ 2,815 more rows</code></pre>
</div>
<p>Selecting one column where several tips are respresented and showing an example of how the nested variable looks like, using eg. nested country variable</p>
<div class="cell">
<pre><code>#&gt; # A tibble: 11 × 3
#&gt;    label            country          pos
#&gt;    &lt;chr&gt;            &lt;chr&gt;          &lt;int&gt;
#&gt;  1 EPI_ISL_13969425 United Kingdom     1
#&gt;  2 EPI_ISL_13969426 United Kingdom     2
#&gt;  3 EPI_ISL_14811896 Sweden             3
#&gt;  4 EPI_ISL_16122154 Germany            4
#&gt;  5 EPI_ISL_14702899 France             5
#&gt;  6 EPI_ISL_15267025 Netherlands        6
#&gt;  7 EPI_ISL_15267027 Netherlands        7
#&gt;  8 EPI_ISL_14573001 Germany            8
#&gt;  9 2022_07_1116_2   Norway             9
#&gt; 10 2022_07_1116_3   Norway            10
#&gt; 11 2022_07_1116_1   Norway            11</code></pre>
</div>
<p><code>pos</code> is the index, in case we need to refer to it. <code>label</code> is the label of the sequences that are represented</p>
<p>Now we can look at the contingency table for the same variable</p>
<div class="cell">
<pre><code>#&gt; # A tibble: 6 × 4
#&gt;   country        `freq_levels@country` `nb_levels@country`   pos
#&gt;   &lt;chr&gt;                          &lt;int&gt;               &lt;int&gt; &lt;int&gt;
#&gt; 1 United Kingdom                     2                   6     1
#&gt; 2 Sweden                             1                   6     2
#&gt; 3 Germany                            2                   6     3
#&gt; 4 France                             1                   6     4
#&gt; 5 Netherlands                        2                   6     5
#&gt; 6 Norway                             3                   6     6</code></pre>
</div>
<p><code>freq_levels@country</code> : is the number of observations in each country <code>nb_levels@country</code> : is common in the nested table, its the number of levels of the variable country</p>
<ul>
<li>NOW: we can reinsert the augmented data to the plot</li>
</ul>
</section>
<section id="filtering-based-on-criteria-based-on-information-from-nested-metadata" class="level3">
<h3 class="anchored" data-anchor-id="filtering-based-on-criteria-based-on-information-from-nested-metadata">3. Filtering based on criteria based on information from nested metadata</h3>
<ul>
<li>Filtering based on criteria that are found in the nested metadata that is attached to the tree plot_data</li>
</ul>
<p>We add a tag - if there is at least one isolate that is represented by a label in the phylogenetic tree has been collected in Norway</p>
<div class="cell">
<pre><code>#&gt;  [1] "parent"                     "node"                      
#&gt;  [3] "branch.length"              "label"                     
#&gt;  [5] "isTip"                      "x"                         
#&gt;  [7] "y"                          "branch"                    
#&gt;  [9] "angle"                      "labels_bootstrap"          
#&gt; [11] "node_depth"                 "descendants_tips_from_node"
#&gt; [13] "isMono_noboot"              "isMono"                    
#&gt; [15] "edge_num"                   "tip_metalink"              
#&gt; [17] "nested_country"             "ctg_country"               
#&gt; [19] "nested_Host"                "ctg_Host"                  
#&gt; [21] "nested_Host_latin"          "ctg_Host_latin"            
#&gt; [23] "nested_year"                "ctg_year"                  
#&gt; [25] "nested_continent"           "ctg_continent"             
#&gt; [27] "nested_Host_family"         "ctg_Host_family"           
#&gt; [29] "inNorway"                   "isWTE"                     
#&gt; [31] "category_highlight"</code></pre>
</div>
</section>
<section id="test-visualisation" class="level3">
<h3 class="anchored" data-anchor-id="test-visualisation">4. Test visualisation</h3>
<p>We can visualize which nodes can be collapsed and at which depths We can try to vary the depth of notes to be collapsed</p>
<div class="cell">
<pre><code>#&gt;  [1] "parent"                     "node"                      
#&gt;  [3] "branch.length"              "label"                     
#&gt;  [5] "isTip"                      "x"                         
#&gt;  [7] "y"                          "branch"                    
#&gt;  [9] "angle"                      "labels_bootstrap"          
#&gt; [11] "node_depth"                 "descendants_tips_from_node"
#&gt; [13] "isMono_noboot"              "isMono"                    
#&gt; [15] "edge_num"                   "tip_metalink"              
#&gt; [17] "nested_country"             "ctg_country"               
#&gt; [19] "nested_Host"                "ctg_Host"                  
#&gt; [21] "nested_Host_latin"          "ctg_Host_latin"            
#&gt; [23] "nested_year"                "ctg_year"                  
#&gt; [25] "nested_continent"           "ctg_continent"             
#&gt; [27] "nested_Host_family"         "ctg_Host_family"           
#&gt; [29] "inNorway"                   "isWTE"                     
#&gt; [31] "category_highlight"</code></pre>
</div>
<p>We add the node labels so we can know which ones we can collapse afterwards</p>
<div class="cell">
<pre><code>#&gt; Warning: There were 1419 warnings in `filter()`.
#&gt; The first warning was:
#&gt; ℹ In argument: `as.numeric(labels_bootstrap) &gt;= 95`.
#&gt; ℹ In row 1.
#&gt; Caused by warning:
#&gt; ! NAs introduced by coercion
#&gt; ℹ Run `dplyr::last_dplyr_warnings()` to see the 1418 remaining warnings.</code></pre>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<pre><code>#&gt; Warning: There were 1419 warnings in `filter()`.
#&gt; The first warning was:
#&gt; ℹ In argument: `as.numeric(labels_bootstrap) &gt;= 95`.
#&gt; ℹ In row 1.
#&gt; Caused by warning:
#&gt; ! NAs introduced by coercion
#&gt; ℹ Run `dplyr::last_dplyr_warnings()` to see the 1418 remaining warnings.</code></pre>
</div>
</section>
<section id="define-criteria-priority-of-nodes-to-collapse" class="level3">
<h3 class="anchored" data-anchor-id="define-criteria-priority-of-nodes-to-collapse">5. Define criteria priority of nodes to collapse</h3>
<ol type="1">
<li>all monophyletic nodes (no need for bootstrap support) where isolates belong to the same countries (supposed to represent same event) AND are not from Norway.</li>
</ol>
<p>We need to extract a list where only one country is represented on node</p>
<p>NB: isMonoNobot is always TRUEE for nodes, as only get descendants from nodes (so I could have removed this but ok)</p>
<ol start="2" type="1">
<li>monophyletic nodes (incl.&nbsp;bootstrap support) where our isolates are not included (visual plot to define limit of depth)</li>
</ol>
<p>Here I choose a limit of node depth =&gt; 5 We will need to collapse nodes that are deepest first (ensure max collapsing, as the following nodes can be subordinate)</p>
<p>By order of priority, the nodes to collapse are:</p>
</section>
</section>
<section id="b.-collapsing-nodes-step-1" class="level2">
<h2 class="anchored" data-anchor-id="b.-collapsing-nodes-step-1">B. Collapsing nodes (step 1)</h2>
<blockquote class="blockquote">
<p>apparte to update data and to get objects easier to work with saving the tree that had node labels</p>
</blockquote>
<ul>
<li>collapsing the nodes recursively. For this we require a list of descendants of for nodes to collapse # NOW</li>
</ul>
<p>Collapsing the nodes of the tree</p>
<div class="cell">
<pre><code>#&gt; recuring collapsing of nodes
#&gt; Node  N785  does not exist in the tree
#&gt; Node  N492  does not exist in the tree
#&gt; Node  N1253  does not exist in the tree
#&gt; Node  N725  does not exist in the tree
#&gt; Node  N1193  does not exist in the tree
#&gt; Node  N1194  does not exist in the tree
#&gt; Node  N1195  does not exist in the tree
#&gt; Node  N1198  does not exist in the tree
#&gt; Node  N1200  does not exist in the tree
#&gt; Node  N788  does not exist in the tree
#&gt; Node  N829  does not exist in the tree
#&gt; Node  N1201  does not exist in the tree
#&gt; Node  N504  does not exist in the tree
#&gt; Node  N1202  does not exist in the tree
#&gt; Node  N1254  does not exist in the tree
#&gt; Node  N505  does not exist in the tree
#&gt; Node  N1203  does not exist in the tree
#&gt; Node  N247  does not exist in the tree
#&gt; Node  N830  does not exist in the tree
#&gt; Node  N1204  does not exist in the tree
#&gt; Node  N506  does not exist in the tree
#&gt; Node  N727  does not exist in the tree
#&gt; Node  N1205  does not exist in the tree
#&gt; Node  N248  does not exist in the tree
#&gt; Node  N249  does not exist in the tree
#&gt; Node  N509  does not exist in the tree
#&gt; Node  N1341  does not exist in the tree
#&gt; Node  N250  does not exist in the tree
#&gt; Node  N731  does not exist in the tree
#&gt; Node  N1342  does not exist in the tree
#&gt; Node  N510  does not exist in the tree
#&gt; Node  N732  does not exist in the tree
#&gt; Node  N1070  does not exist in the tree
#&gt; Node  N1343  does not exist in the tree
#&gt; Node  N1071  does not exist in the tree
#&gt; Node  N734  does not exist in the tree
#&gt; Node  N1206  does not exist in the tree
#&gt; Node  N557  does not exist in the tree
#&gt; Node  N735  does not exist in the tree
#&gt; Node  N800  does not exist in the tree
#&gt; Node  N1207  does not exist in the tree
#&gt; Node  N511  does not exist in the tree
#&gt; Node  N558  does not exist in the tree
#&gt; Node  N831  does not exist in the tree
#&gt; Node  N1073  does not exist in the tree
#&gt; Node  N252  does not exist in the tree
#&gt; Node  N447  does not exist in the tree
#&gt; Node  N559  does not exist in the tree
#&gt; Node  N832  does not exist in the tree
#&gt; Node  N1074  does not exist in the tree
#&gt; Node  N448  does not exist in the tree
#&gt; Node  N736  does not exist in the tree
#&gt; Node  N833  does not exist in the tree
#&gt; Node  N449  does not exist in the tree
#&gt; Node  N737  does not exist in the tree
#&gt; Node  N1075  does not exist in the tree
#&gt; Node  N1208  does not exist in the tree
#&gt; Node  N1344  does not exist in the tree
#&gt; Node  N561  does not exist in the tree
#&gt; Node  N1209  does not exist in the tree
#&gt; Node  N203  does not exist in the tree
#&gt; Node  N512  does not exist in the tree
#&gt; Node  N562  does not exist in the tree
#&gt; Node  N738  does not exist in the tree
#&gt; Node  N878  does not exist in the tree
#&gt; Node  N1210  does not exist in the tree
#&gt; Node  N129  does not exist in the tree
#&gt; Node  N204  does not exist in the tree
#&gt; Node  N513  does not exist in the tree
#&gt; Node  N563  does not exist in the tree
#&gt; Node  N834  does not exist in the tree
#&gt; Node  N514  does not exist in the tree
#&gt; Node  N1273  does not exist in the tree
#&gt; Node  N99  does not exist in the tree
#&gt; Node  N420  does not exist in the tree
#&gt; Node  N741  does not exist in the tree
#&gt; Node  N880  does not exist in the tree
#&gt; Node  N1308  does not exist in the tree
#&gt; Node  N253  does not exist in the tree
#&gt; Node  N421  does not exist in the tree
#&gt; Node  N564  does not exist in the tree
#&gt; Node  N810  does not exist in the tree
#&gt; Node  N881  does not exist in the tree
#&gt; Node  N1274  does not exist in the tree
#&gt; Node  N254  does not exist in the tree
#&gt; Node  N422  does not exist in the tree
#&gt; Node  N835  does not exist in the tree
#&gt; Node  N882  does not exist in the tree
#&gt; Node  N902  does not exist in the tree
#&gt; Node  N1255  does not exist in the tree
#&gt; Node  N423  does not exist in the tree
#&gt; Node  N565  does not exist in the tree
#&gt; Node  N678  does not exist in the tree
#&gt; Node  N903  does not exist in the tree
#&gt; Node  N404  does not exist in the tree
#&gt; Node  N424  does not exist in the tree
#&gt; Node  N566  does not exist in the tree
#&gt; Node  N1217  does not exist in the tree
#&gt; Node  N104  does not exist in the tree
#&gt; Node  N405  does not exist in the tree
#&gt; Node  N425  does not exist in the tree
#&gt; Node  N567  does not exist in the tree
#&gt; Node  N764  does not exist in the tree
#&gt; Node  N1141  does not exist in the tree
#&gt; Node  N1218  does not exist in the tree
#&gt; Node  N1275  does not exist in the tree
#&gt; Node  N206  does not exist in the tree
#&gt; Node  N258  does not exist in the tree
#&gt; Node  N680  does not exist in the tree
#&gt; Node  N1009  does not exist in the tree
#&gt; Node  N1219  does not exist in the tree
#&gt; Node  N1276  does not exist in the tree
#&gt; Node  N1294  does not exist in the tree
#&gt; Node  N1386  does not exist in the tree
#&gt; Node  N651  does not exist in the tree
#&gt; Node  N765  does not exist in the tree
#&gt; Node  N836  does not exist in the tree
#&gt; Node  N1010  does not exist in the tree
#&gt; Node  N1077  does not exist in the tree
#&gt; Node  N1220  does not exist in the tree
#&gt; Node  N1295  does not exist in the tree
#&gt; Node  N1387  does not exist in the tree
#&gt; Node  N105  does not exist in the tree
#&gt; Node  N208  does not exist in the tree
#&gt; Node  N426  does not exist in the tree
#&gt; Node  N450  does not exist in the tree
#&gt; Node  N462  does not exist in the tree
#&gt; Node  N766  does not exist in the tree
#&gt; Node  N920  does not exist in the tree
#&gt; Node  N1078  does not exist in the tree
#&gt; Node  N1316  does not exist in the tree
#&gt; Node  N1329  does not exist in the tree
#&gt; Node  N1345  does not exist in the tree
#&gt; Node  N1357  does not exist in the tree
#&gt; Node  N1388  does not exist in the tree
#&gt; Node  N106  does not exist in the tree
#&gt; Node  N233  does not exist in the tree
#&gt; Node  N261  does not exist in the tree
#&gt; Node  N610  does not exist in the tree
#&gt; Node  N652  does not exist in the tree
#&gt; Node  N683  does not exist in the tree
#&gt; Node  N789  does not exist in the tree
#&gt; Node  N811  does not exist in the tree
#&gt; Node  N921  does not exist in the tree
#&gt; Node  N1079  does not exist in the tree
#&gt; Node  N1277  does not exist in the tree
#&gt; Node  N1317  does not exist in the tree
#&gt; Node  N1346  does not exist in the tree
#&gt; Node  N463  does not exist in the tree
#&gt; Node  N515  does not exist in the tree
#&gt; Node  N525  does not exist in the tree
#&gt; Node  N611  does not exist in the tree
#&gt; Node  N743  does not exist in the tree
#&gt; Node  N790  does not exist in the tree
#&gt; Node  N910  does not exist in the tree
#&gt; Node  N988  does not exist in the tree
#&gt; Node  N1039  does not exist in the tree
#&gt; Node  N1263  does not exist in the tree
#&gt; Node  N107  does not exist in the tree
#&gt; Node  N234  does not exist in the tree
#&gt; Node  N310  does not exist in the tree
#&gt; Node  N495  does not exist in the tree
#&gt; Node  N526  does not exist in the tree
#&gt; Node  N684  does not exist in the tree
#&gt; Node  N744  does not exist in the tree
#&gt; Node  N767  does not exist in the tree
#&gt; Node  N801  does not exist in the tree
#&gt; Node  N812  does not exist in the tree
#&gt; Node  N859  does not exist in the tree
#&gt; Node  N1040  does not exist in the tree
#&gt; Node  N1299  does not exist in the tree
#&gt; Node  N1358  does not exist in the tree
#&gt; Node  N1389  does not exist in the tree
#&gt; Node  N108  does not exist in the tree
#&gt; Node  N132  does not exist in the tree
#&gt; Node  N407  does not exist in the tree
#&gt; Node  N768  does not exist in the tree
#&gt; Node  N802  does not exist in the tree
#&gt; Node  N837  does not exist in the tree
#&gt; Node  N911  does not exist in the tree
#&gt; Node  N924  does not exist in the tree
#&gt; Node  N998  does not exist in the tree
#&gt; Node  N1164  does not exist in the tree
#&gt; Node  N1264  does not exist in the tree
#&gt; Node  N1300  does not exist in the tree
#&gt; Node  N1318  does not exist in the tree
#&gt; Node  N1330  does not exist in the tree
#&gt; Node  N133  does not exist in the tree
#&gt; Node  N262  does not exist in the tree
#&gt; Node  N272  does not exist in the tree
#&gt; Node  N527  does not exist in the tree
#&gt; Node  N653  does not exist in the tree
#&gt; Node  N665  does not exist in the tree
#&gt; Node  N686  does not exist in the tree
#&gt; Node  N746  does not exist in the tree
#&gt; Node  N769  does not exist in the tree
#&gt; Node  N793  does not exist in the tree
#&gt; Node  N822  does not exist in the tree
#&gt; Node  N884  does not exist in the tree
#&gt; Node  N912  does not exist in the tree
#&gt; Node  N925  does not exist in the tree
#&gt; Node  N999  does not exist in the tree
#&gt; Node  N1012  does not exist in the tree
#&gt; Node  N1107  does not exist in the tree
#&gt; Node  N1143  does not exist in the tree
#&gt; Node  N1165  does not exist in the tree
#&gt; Node  N1181  does not exist in the tree
#&gt; Node  N1221  does not exist in the tree
#&gt; Node  N1256  does not exist in the tree
#&gt; Node  N1265  does not exist in the tree
#&gt; Node  N1309  does not exist in the tree
#&gt; Node  N1391  does not exist in the tree
#&gt; Node  N134  does not exist in the tree
#&gt; Node  N210  does not exist in the tree
#&gt; Node  N220  does not exist in the tree
#&gt; Node  N263  does not exist in the tree
#&gt; Node  N312  does not exist in the tree
#&gt; Node  N408  does not exist in the tree
#&gt; Node  N451  does not exist in the tree
#&gt; Node  N474  does not exist in the tree
#&gt; Node  N519  does not exist in the tree
#&gt; Node  N528  does not exist in the tree
#&gt; Node  N574  does not exist in the tree
#&gt; Node  N639  does not exist in the tree
#&gt; Node  N654  does not exist in the tree
#&gt; Node  N747  does not exist in the tree
#&gt; Node  N794  does not exist in the tree
#&gt; Node  N813  does not exist in the tree
#&gt; Node  N870  does not exist in the tree
#&gt; Node  N885  does not exist in the tree
#&gt; Node  N904  does not exist in the tree
#&gt; Node  N913  does not exist in the tree
#&gt; Node  N926  does not exist in the tree
#&gt; Node  N980  does not exist in the tree
#&gt; Node  N1023  does not exist in the tree
#&gt; Node  N1080  does not exist in the tree
#&gt; Node  N1090  does not exist in the tree
#&gt; Node  N1144  does not exist in the tree
#&gt; Node  N1150  does not exist in the tree
#&gt; Node  N1166  does not exist in the tree
#&gt; Node  N1182  does not exist in the tree
#&gt; Node  N1211  does not exist in the tree
#&gt; Node  N1222  does not exist in the tree
#&gt; Node  N1236  does not exist in the tree
#&gt; Node  N1257  does not exist in the tree
#&gt; Node  N1266  does not exist in the tree
#&gt; Node  N1310  does not exist in the tree
#&gt; Node  N1351  does not exist in the tree
#&gt; Node  N1369  does not exist in the tree
#&gt; Node  N135  does not exist in the tree
#&gt; Node  N221  does not exist in the tree
#&gt; Node  N235  does not exist in the tree
#&gt; Node  N313  does not exist in the tree
#&gt; Node  N428  does not exist in the tree
#&gt; Node  N457  does not exist in the tree
#&gt; Node  N464  does not exist in the tree
#&gt; Node  N499  does not exist in the tree
#&gt; Node  N520  does not exist in the tree
#&gt; Node  N569  does not exist in the tree
#&gt; Node  N599  does not exist in the tree
#&gt; Node  N604  does not exist in the tree
#&gt; Node  N613  does not exist in the tree
#&gt; Node  N640  does not exist in the tree
#&gt; Node  N646  does not exist in the tree
#&gt; Node  N710  does not exist in the tree
#&gt; Node  N748  does not exist in the tree
#&gt; Node  N753  does not exist in the tree
#&gt; Node  N770  does not exist in the tree
#&gt; Node  N805  does not exist in the tree
#&gt; Node  N840  does not exist in the tree
#&gt; Node  N871  does not exist in the tree
#&gt; Node  N886  does not exist in the tree
#&gt; Node  N895  does not exist in the tree
#&gt; Node  N905  does not exist in the tree
#&gt; Node  N914  does not exist in the tree
#&gt; Node  N968  does not exist in the tree
#&gt; Node  N981  does not exist in the tree
#&gt; Node  N1000  does not exist in the tree
#&gt; Node  N1042  does not exist in the tree
#&gt; Node  N1091  does not exist in the tree
#&gt; Node  N1108  does not exist in the tree
#&gt; Node  N1145  does not exist in the tree
#&gt; Node  N1151  does not exist in the tree
#&gt; Node  N1183  does not exist in the tree
#&gt; Node  N1212  does not exist in the tree
#&gt; Node  N1223  does not exist in the tree
#&gt; Node  N1228  does not exist in the tree
#&gt; Node  N1258  does not exist in the tree
#&gt; Node  N1278  does not exist in the tree
#&gt; Node  N1283  does not exist in the tree
#&gt; Node  N1303  does not exist in the tree
#&gt; Node  N1352  does not exist in the tree
#&gt; Node  N1359  does not exist in the tree
#&gt; Node  N1370  does not exist in the tree
#&gt; Node  N1379  does not exist in the tree
#&gt; Node  N1400  does not exist in the tree
#&gt; Node  N61  does not exist in the tree
#&gt; Node  N112  does not exist in the tree
#&gt; Node  N136  does not exist in the tree
#&gt; Node  N143  does not exist in the tree
#&gt; Node  N216  does not exist in the tree
#&gt; Node  N222  does not exist in the tree
#&gt; Node  N236  does not exist in the tree
#&gt; Node  N265  does not exist in the tree
#&gt; Node  N279  does not exist in the tree
#&gt; Node  N286  does not exist in the tree
#&gt; Node  N314  does not exist in the tree
#&gt; Node  N332  does not exist in the tree
#&gt; Node  N366  does not exist in the tree
#&gt; Node  N409  does not exist in the tree
#&gt; Node  N429  does not exist in the tree
#&gt; Node  N433  does not exist in the tree
#&gt; Node  N452  does not exist in the tree
#&gt; Node  N458  does not exist in the tree
#&gt; Node  N465  does not exist in the tree
#&gt; Node  N469  does not exist in the tree
#&gt; Node  N500  does not exist in the tree
#&gt; Node  N521  does not exist in the tree
#&gt; Node  N529  does not exist in the tree
#&gt; Node  N535  does not exist in the tree
#&gt; Node  N547  does not exist in the tree
#&gt; Node  N575  does not exist in the tree
#&gt; Node  N605  does not exist in the tree
#&gt; Node  N614  does not exist in the tree
#&gt; Node  N655  does not exist in the tree
#&gt; Node  N666  does not exist in the tree
#&gt; Node  N711  does not exist in the tree
#&gt; Node  N754  does not exist in the tree
#&gt; Node  N771  does not exist in the tree
#&gt; Node  N795  does not exist in the tree
#&gt; Node  N806  does not exist in the tree
#&gt; Node  N823  does not exist in the tree
#&gt; Node  N841  does not exist in the tree
#&gt; Node  N845  does not exist in the tree
#&gt; Node  N849  does not exist in the tree
#&gt; Node  N860  does not exist in the tree
#&gt; Node  N864  does not exist in the tree
#&gt; Node  N872  does not exist in the tree
#&gt; Node  N891  does not exist in the tree
#&gt; Node  N896  does not exist in the tree
#&gt; Node  N906  does not exist in the tree
#&gt; Node  N915  does not exist in the tree
#&gt; Node  N969  does not exist in the tree
#&gt; Node  N982  does not exist in the tree
#&gt; Node  N993  does not exist in the tree
#&gt; Node  N1001  does not exist in the tree
#&gt; Node  N1013  does not exist in the tree
#&gt; Node  N1019  does not exist in the tree
#&gt; Node  N1086  does not exist in the tree
#&gt; Node  N1092  does not exist in the tree
#&gt; Node  N1096  does not exist in the tree
#&gt; Node  N1109  does not exist in the tree
#&gt; Node  N1129  does not exist in the tree
#&gt; Node  N1146  does not exist in the tree
#&gt; Node  N1152  does not exist in the tree
#&gt; Node  N1184  does not exist in the tree
#&gt; Node  N1213  does not exist in the tree
#&gt; Node  N1229  does not exist in the tree
#&gt; Node  N1242  does not exist in the tree
#&gt; Node  N1267  does not exist in the tree
#&gt; Node  N1279  does not exist in the tree
#&gt; Node  N1304  does not exist in the tree
#&gt; Node  N1319  does not exist in the tree
#&gt; Node  N1331  does not exist in the tree
#&gt; Node  N1347  does not exist in the tree
#&gt; Node  N1360  does not exist in the tree
#&gt; Node  N1375  does not exist in the tree
#&gt; Node  N57  does not exist in the tree
#&gt; Node  N59  does not exist in the tree
#&gt; Node  N60  does not exist in the tree
#&gt; Node  N64  does not exist in the tree
#&gt; Node  N116  does not exist in the tree
#&gt; Node  N118  does not exist in the tree
#&gt; Node  N131  does not exist in the tree
#&gt; Node  N137  does not exist in the tree
#&gt; Node  N138  does not exist in the tree
#&gt; Node  N139  does not exist in the tree
#&gt; Node  N141  does not exist in the tree
#&gt; Node  N144  does not exist in the tree
#&gt; Node  N145  does not exist in the tree
#&gt; Node  N146  does not exist in the tree
#&gt; Node  N155  does not exist in the tree
#&gt; Node  N156  does not exist in the tree
#&gt; Node  N170  does not exist in the tree
#&gt; Node  N171  does not exist in the tree
#&gt; Node  N174  does not exist in the tree
#&gt; Node  N175  does not exist in the tree
#&gt; Node  N180  does not exist in the tree
#&gt; Node  N183  does not exist in the tree
#&gt; Node  N184  does not exist in the tree
#&gt; Node  N192  does not exist in the tree
#&gt; Node  N193  does not exist in the tree
#&gt; Node  N195  does not exist in the tree
#&gt; Node  N212  does not exist in the tree
#&gt; Node  N215  does not exist in the tree
#&gt; Node  N219  does not exist in the tree
#&gt; Node  N223  does not exist in the tree
#&gt; Node  N224  does not exist in the tree
#&gt; Node  N225  does not exist in the tree
#&gt; Node  N240  does not exist in the tree
#&gt; Node  N241  does not exist in the tree
#&gt; Node  N242  does not exist in the tree
#&gt; Node  N246  does not exist in the tree
#&gt; Node  N257  does not exist in the tree
#&gt; Node  N264  does not exist in the tree
#&gt; Node  N266  does not exist in the tree
#&gt; Node  N267  does not exist in the tree
#&gt; Node  N268  does not exist in the tree
#&gt; Node  N270  does not exist in the tree
#&gt; Node  N271  does not exist in the tree
#&gt; Node  N277  does not exist in the tree
#&gt; Node  N278  does not exist in the tree
#&gt; Node  N281  does not exist in the tree
#&gt; Node  N282  does not exist in the tree
#&gt; Node  N284  does not exist in the tree
#&gt; Node  N291  does not exist in the tree
#&gt; Node  N297  does not exist in the tree
#&gt; Node  N299  does not exist in the tree
#&gt; Node  N304  does not exist in the tree
#&gt; Node  N305  does not exist in the tree
#&gt; Node  N306  does not exist in the tree
#&gt; Node  N315  does not exist in the tree
#&gt; Node  N316  does not exist in the tree
#&gt; Node  N317  does not exist in the tree
#&gt; Node  N318  does not exist in the tree
#&gt; Node  N335  does not exist in the tree
#&gt; Node  N340  does not exist in the tree
#&gt; Node  N391  does not exist in the tree
#&gt; Node  N392  does not exist in the tree
#&gt; Node  N393  does not exist in the tree
#&gt; Node  N396  does not exist in the tree
#&gt; Node  N397  does not exist in the tree
#&gt; Node  N398  does not exist in the tree
#&gt; Node  N399  does not exist in the tree
#&gt; Node  N412  does not exist in the tree
#&gt; Node  N413  does not exist in the tree
#&gt; Node  N414  does not exist in the tree
#&gt; Node  N415  does not exist in the tree
#&gt; Node  N416  does not exist in the tree
#&gt; Node  N417  does not exist in the tree
#&gt; Node  N418  does not exist in the tree
#&gt; Node  N419  does not exist in the tree
#&gt; Node  N430  does not exist in the tree
#&gt; Node  N431  does not exist in the tree
#&gt; Node  N432  does not exist in the tree
#&gt; Node  N434  does not exist in the tree
#&gt; Node  N435  does not exist in the tree
#&gt; Node  N436  does not exist in the tree
#&gt; Node  N437  does not exist in the tree
#&gt; Node  N438  does not exist in the tree
#&gt; Node  N439  does not exist in the tree
#&gt; Node  N440  does not exist in the tree
#&gt; Node  N441  does not exist in the tree
#&gt; Node  N442  does not exist in the tree
#&gt; Node  N466  does not exist in the tree
#&gt; Node  N467  does not exist in the tree
#&gt; Node  N468  does not exist in the tree
#&gt; Node  N470  does not exist in the tree
#&gt; Node  N471  does not exist in the tree
#&gt; Node  N476  does not exist in the tree
#&gt; Node  N480  does not exist in the tree
#&gt; Node  N481  does not exist in the tree
#&gt; Node  N482  does not exist in the tree
#&gt; Node  N483  does not exist in the tree
#&gt; Node  N497  does not exist in the tree
#&gt; Node  N501  does not exist in the tree
#&gt; Node  N502  does not exist in the tree
#&gt; Node  N503  does not exist in the tree
#&gt; Node  N507  does not exist in the tree
#&gt; Node  N508  does not exist in the tree
#&gt; Node  N516  does not exist in the tree
#&gt; Node  N517  does not exist in the tree
#&gt; Node  N518  does not exist in the tree
#&gt; Node  N522  does not exist in the tree
#&gt; Node  N523  does not exist in the tree
#&gt; Node  N524  does not exist in the tree
#&gt; Node  N530  does not exist in the tree
#&gt; Node  N531  does not exist in the tree
#&gt; Node  N532  does not exist in the tree
#&gt; Node  N533  does not exist in the tree
#&gt; Node  N534  does not exist in the tree
#&gt; Node  N536  does not exist in the tree
#&gt; Node  N537  does not exist in the tree
#&gt; Node  N538  does not exist in the tree
#&gt; Node  N539  does not exist in the tree
#&gt; Node  N540  does not exist in the tree
#&gt; Node  N541  does not exist in the tree
#&gt; Node  N542  does not exist in the tree
#&gt; Node  N543  does not exist in the tree
#&gt; Node  N545  does not exist in the tree
#&gt; Node  N546  does not exist in the tree
#&gt; Node  N549  does not exist in the tree
#&gt; Node  N550  does not exist in the tree
#&gt; Node  N553  does not exist in the tree
#&gt; Node  N554  does not exist in the tree
#&gt; Node  N571  does not exist in the tree
#&gt; Node  N578  does not exist in the tree
#&gt; Node  N580  does not exist in the tree
#&gt; Node  N581  does not exist in the tree
#&gt; Node  N582  does not exist in the tree
#&gt; Node  N585  does not exist in the tree
#&gt; Node  N590  does not exist in the tree
#&gt; Node  N591  does not exist in the tree
#&gt; Node  N596  does not exist in the tree
#&gt; Node  N597  does not exist in the tree
#&gt; Node  N600  does not exist in the tree
#&gt; Node  N601  does not exist in the tree
#&gt; Node  N606  does not exist in the tree
#&gt; Node  N607  does not exist in the tree
#&gt; Node  N608  does not exist in the tree
#&gt; Node  N615  does not exist in the tree
#&gt; Node  N616  does not exist in the tree
#&gt; Node  N617  does not exist in the tree
#&gt; Node  N618  does not exist in the tree
#&gt; Node  N619  does not exist in the tree
#&gt; Node  N620  does not exist in the tree
#&gt; Node  N623  does not exist in the tree
#&gt; Node  N624  does not exist in the tree
#&gt; Node  N625  does not exist in the tree
#&gt; Node  N644  does not exist in the tree
#&gt; Node  N648  does not exist in the tree
#&gt; Node  N649  does not exist in the tree
#&gt; Node  N659  does not exist in the tree
#&gt; Node  N660  does not exist in the tree
#&gt; Node  N661  does not exist in the tree
#&gt; Node  N662  does not exist in the tree
#&gt; Node  N668  does not exist in the tree
#&gt; Node  N669  does not exist in the tree
#&gt; Node  N681  does not exist in the tree
#&gt; Node  N682  does not exist in the tree
#&gt; Node  N685  does not exist in the tree
#&gt; Node  N687  does not exist in the tree
#&gt; Node  N688  does not exist in the tree
#&gt; Node  N689  does not exist in the tree
#&gt; Node  N690  does not exist in the tree
#&gt; Node  N691  does not exist in the tree
#&gt; Node  N692  does not exist in the tree
#&gt; Node  N693  does not exist in the tree
#&gt; Node  N703  does not exist in the tree
#&gt; Node  N712  does not exist in the tree
#&gt; Node  N713  does not exist in the tree
#&gt; Node  N714  does not exist in the tree
#&gt; Node  N728  does not exist in the tree
#&gt; Node  N729  does not exist in the tree
#&gt; Node  N730  does not exist in the tree
#&gt; Node  N739  does not exist in the tree
#&gt; Node  N740  does not exist in the tree
#&gt; Node  N745  does not exist in the tree
#&gt; Node  N749  does not exist in the tree
#&gt; Node  N750  does not exist in the tree
#&gt; Node  N751  does not exist in the tree
#&gt; Node  N752  does not exist in the tree
#&gt; Node  N756  does not exist in the tree
#&gt; Node  N757  does not exist in the tree
#&gt; Node  N758  does not exist in the tree
#&gt; Node  N759  does not exist in the tree
#&gt; Node  N760  does not exist in the tree
#&gt; Node  N761  does not exist in the tree
#&gt; Node  N763  does not exist in the tree
#&gt; Node  N775  does not exist in the tree
#&gt; Node  N776  does not exist in the tree
#&gt; Node  N777  does not exist in the tree
#&gt; Node  N778  does not exist in the tree
#&gt; Node  N780  does not exist in the tree
#&gt; Node  N791  does not exist in the tree
#&gt; Node  N792  does not exist in the tree
#&gt; Node  N796  does not exist in the tree
#&gt; Node  N797  does not exist in the tree
#&gt; Node  N798  does not exist in the tree
#&gt; Node  N799  does not exist in the tree
#&gt; Node  N803  does not exist in the tree
#&gt; Node  N804  does not exist in the tree
#&gt; Node  N815  does not exist in the tree
#&gt; Node  N816  does not exist in the tree
#&gt; Node  N817  does not exist in the tree
#&gt; Node  N818  does not exist in the tree
#&gt; Node  N820  does not exist in the tree
#&gt; Node  N821  does not exist in the tree
#&gt; Node  N824  does not exist in the tree
#&gt; Node  N825  does not exist in the tree
#&gt; Node  N826  does not exist in the tree
#&gt; Node  N828  does not exist in the tree
#&gt; Node  N838  does not exist in the tree
#&gt; Node  N839  does not exist in the tree
#&gt; Node  N842  does not exist in the tree
#&gt; Node  N843  does not exist in the tree
#&gt; Node  N844  does not exist in the tree
#&gt; Node  N848  does not exist in the tree
#&gt; Node  N851  does not exist in the tree
#&gt; Node  N852  does not exist in the tree
#&gt; Node  N853  does not exist in the tree
#&gt; Node  N854  does not exist in the tree
#&gt; Node  N855  does not exist in the tree
#&gt; Node  N856  does not exist in the tree
#&gt; Node  N857  does not exist in the tree
#&gt; Node  N858  does not exist in the tree
#&gt; Node  N861  does not exist in the tree
#&gt; Node  N862  does not exist in the tree
#&gt; Node  N863  does not exist in the tree
#&gt; Node  N866  does not exist in the tree
#&gt; Node  N867  does not exist in the tree
#&gt; Node  N873  does not exist in the tree
#&gt; Node  N874  does not exist in the tree
#&gt; Node  N875  does not exist in the tree
#&gt; Node  N892  does not exist in the tree
#&gt; Node  N893  does not exist in the tree
#&gt; Node  N897  does not exist in the tree
#&gt; Node  N898  does not exist in the tree
#&gt; Node  N899  does not exist in the tree
#&gt; Node  N909  does not exist in the tree
#&gt; Node  N928  does not exist in the tree
#&gt; Node  N929  does not exist in the tree
#&gt; Node  N930  does not exist in the tree
#&gt; Node  N931  does not exist in the tree
#&gt; Node  N954  does not exist in the tree
#&gt; Node  N955  does not exist in the tree
#&gt; Node  N956  does not exist in the tree
#&gt; Node  N959  does not exist in the tree
#&gt; Node  N960  does not exist in the tree
#&gt; Node  N961  does not exist in the tree
#&gt; Node  N962  does not exist in the tree
#&gt; Node  N964  does not exist in the tree
#&gt; Node  N965  does not exist in the tree
#&gt; Node  N967  does not exist in the tree
#&gt; Node  N970  does not exist in the tree
#&gt; Node  N971  does not exist in the tree
#&gt; Node  N972  does not exist in the tree
#&gt; Node  N976  does not exist in the tree
#&gt; Node  N977  does not exist in the tree
#&gt; Node  N983  does not exist in the tree
#&gt; Node  N984  does not exist in the tree
#&gt; Node  N985  does not exist in the tree
#&gt; Node  N991  does not exist in the tree
#&gt; Node  N992  does not exist in the tree
#&gt; Node  N994  does not exist in the tree
#&gt; Node  N995  does not exist in the tree
#&gt; Node  N996  does not exist in the tree
#&gt; Node  N997  does not exist in the tree
#&gt; Node  N1002  does not exist in the tree
#&gt; Node  N1003  does not exist in the tree
#&gt; Node  N1004  does not exist in the tree
#&gt; Node  N1005  does not exist in the tree
#&gt; Node  N1007  does not exist in the tree
#&gt; Node  N1021  does not exist in the tree
#&gt; Node  N1022  does not exist in the tree
#&gt; Node  N1025  does not exist in the tree
#&gt; Node  N1028  does not exist in the tree
#&gt; Node  N1035  does not exist in the tree
#&gt; Node  N1043  does not exist in the tree
#&gt; Node  N1044  does not exist in the tree
#&gt; Node  N1045  does not exist in the tree
#&gt; Node  N1046  does not exist in the tree
#&gt; Node  N1047  does not exist in the tree
#&gt; Node  N1048  does not exist in the tree
#&gt; Node  N1050  does not exist in the tree
#&gt; Node  N1051  does not exist in the tree
#&gt; Node  N1052  does not exist in the tree
#&gt; Node  N1081  does not exist in the tree
#&gt; Node  N1082  does not exist in the tree
#&gt; Node  N1083  does not exist in the tree
#&gt; Node  N1084  does not exist in the tree
#&gt; Node  N1085  does not exist in the tree
#&gt; Node  N1087  does not exist in the tree
#&gt; Node  N1088  does not exist in the tree
#&gt; Node  N1089  does not exist in the tree
#&gt; Node  N1093  does not exist in the tree
#&gt; Node  N1094  does not exist in the tree
#&gt; Node  N1095  does not exist in the tree
#&gt; Node  N1100  does not exist in the tree
#&gt; Node  N1101  does not exist in the tree
#&gt; Node  N1114  does not exist in the tree
#&gt; Node  N1115  does not exist in the tree
#&gt; Node  N1116  does not exist in the tree
#&gt; Node  N1117  does not exist in the tree
#&gt; Node  N1118  does not exist in the tree
#&gt; Node  N1122  does not exist in the tree
#&gt; Node  N1128  does not exist in the tree
#&gt; Node  N1130  does not exist in the tree
#&gt; Node  N1131  does not exist in the tree
#&gt; Node  N1132  does not exist in the tree
#&gt; Node  N1134  does not exist in the tree
#&gt; Node  N1140  does not exist in the tree
#&gt; Node  N1147  does not exist in the tree
#&gt; Node  N1148  does not exist in the tree
#&gt; Node  N1149  does not exist in the tree
#&gt; Node  N1153  does not exist in the tree
#&gt; Node  N1154  does not exist in the tree
#&gt; Node  N1155  does not exist in the tree
#&gt; Node  N1170  does not exist in the tree
#&gt; Node  N1171  does not exist in the tree
#&gt; Node  N1173  does not exist in the tree
#&gt; Node  N1186  does not exist in the tree
#&gt; Node  N1187  does not exist in the tree
#&gt; Node  N1189  does not exist in the tree
#&gt; Node  N1190  does not exist in the tree
#&gt; Node  N1214  does not exist in the tree
#&gt; Node  N1215  does not exist in the tree
#&gt; Node  N1216  does not exist in the tree
#&gt; Node  N1227  does not exist in the tree
#&gt; Node  N1230  does not exist in the tree
#&gt; Node  N1231  does not exist in the tree
#&gt; Node  N1232  does not exist in the tree
#&gt; Node  N1233  does not exist in the tree
#&gt; Node  N1234  does not exist in the tree
#&gt; Node  N1235  does not exist in the tree
#&gt; Node  N1238  does not exist in the tree
#&gt; Node  N1239  does not exist in the tree
#&gt; Node  N1245  does not exist in the tree
#&gt; Node  N1247  does not exist in the tree
#&gt; Node  N1259  does not exist in the tree
#&gt; Node  N1260  does not exist in the tree
#&gt; Node  N1261  does not exist in the tree
#&gt; Node  N1262  does not exist in the tree
#&gt; Node  N1268  does not exist in the tree
#&gt; Node  N1269  does not exist in the tree
#&gt; Node  N1270  does not exist in the tree
#&gt; Node  N1271  does not exist in the tree
#&gt; Node  N1272  does not exist in the tree
#&gt; Node  N1282  does not exist in the tree
#&gt; Node  N1287  does not exist in the tree
#&gt; Node  N1288  does not exist in the tree
#&gt; Node  N1289  does not exist in the tree
#&gt; Node  N1290  does not exist in the tree
#&gt; Node  N1291  does not exist in the tree
#&gt; Node  N1292  does not exist in the tree
#&gt; Node  N1293  does not exist in the tree
#&gt; Node  N1296  does not exist in the tree
#&gt; Node  N1297  does not exist in the tree
#&gt; Node  N1298  does not exist in the tree
#&gt; Node  N1301  does not exist in the tree
#&gt; Node  N1302  does not exist in the tree
#&gt; Node  N1305  does not exist in the tree
#&gt; Node  N1306  does not exist in the tree
#&gt; Node  N1307  does not exist in the tree
#&gt; Node  N1315  does not exist in the tree
#&gt; Node  N1320  does not exist in the tree
#&gt; Node  N1321  does not exist in the tree
#&gt; Node  N1322  does not exist in the tree
#&gt; Node  N1323  does not exist in the tree
#&gt; Node  N1324  does not exist in the tree
#&gt; Node  N1325  does not exist in the tree
#&gt; Node  N1332  does not exist in the tree
#&gt; Node  N1333  does not exist in the tree
#&gt; Node  N1334  does not exist in the tree
#&gt; Node  N1335  does not exist in the tree
#&gt; Node  N1336  does not exist in the tree
#&gt; Node  N1337  does not exist in the tree
#&gt; Node  N1338  does not exist in the tree
#&gt; Node  N1339  does not exist in the tree
#&gt; Node  N1340  does not exist in the tree
#&gt; Node  N1348  does not exist in the tree
#&gt; Node  N1349  does not exist in the tree
#&gt; Node  N1350  does not exist in the tree
#&gt; Node  N1353  does not exist in the tree
#&gt; Node  N1354  does not exist in the tree
#&gt; Node  N1355  does not exist in the tree
#&gt; Node  N1356  does not exist in the tree
#&gt; Node  N1361  does not exist in the tree
#&gt; Node  N1362  does not exist in the tree
#&gt; Node  N1363  does not exist in the tree
#&gt; Node  N1364  does not exist in the tree
#&gt; Node  N1365  does not exist in the tree
#&gt; Node  N1366  does not exist in the tree
#&gt; Node  N1367  does not exist in the tree
#&gt; Node  N1368  does not exist in the tree
#&gt; Node  N1371  does not exist in the tree
#&gt; Node  N1372  does not exist in the tree
#&gt; Node  N1373  does not exist in the tree
#&gt; Node  N1374  does not exist in the tree
#&gt; Node  N1377  does not exist in the tree
#&gt; Node  N1378  does not exist in the tree
#&gt; Node  N1382  does not exist in the tree
#&gt; Node  N1383  does not exist in the tree
#&gt; Node  N1390  does not exist in the tree
#&gt; Node  N1392  does not exist in the tree
#&gt; Node  N1393  does not exist in the tree
#&gt; Node  N1394  does not exist in the tree
#&gt; Node  N1398  does not exist in the tree
#&gt; Node  N1399  does not exist in the tree
#&gt; Node  N1402  does not exist in the tree
#&gt; Node  N1404  does not exist in the tree
#&gt; Node  N1406  does not exist in the tree
#&gt; returning the collapsed tree</code></pre>
</div>
<p>Visualization of the collapsed tree</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>we need to merge the metadata with the collapsed data</li>
</ul>
<p>Because we have labelled nodes with name, we can reuse the previous labels</p>
<div class="cell">
<pre><code>#&gt;  [1] "parent"                     "node"                      
#&gt;  [3] "branch.length"              "label"                     
#&gt;  [5] "isTip"                      "x"                         
#&gt;  [7] "y"                          "branch"                    
#&gt;  [9] "angle"                      "labels_bootstrap"          
#&gt; [11] "node_depth"                 "descendants_tips_from_node"
#&gt; [13] "isMono_noboot"              "isMono"                    
#&gt; [15] "edge_num"                   "tip_metalink"              
#&gt; [17] "nested_country"             "ctg_country"               
#&gt; [19] "nested_Host"                "ctg_Host"                  
#&gt; [21] "nested_Host_latin"          "ctg_Host_latin"            
#&gt; [23] "nested_year"                "ctg_year"                  
#&gt; [25] "nested_continent"           "ctg_continent"             
#&gt; [27] "nested_Host_family"         "ctg_Host_family"           
#&gt; [29] "inNorway"                   "isWTE"                     
#&gt; [31] "category_highlight"         "isOneCountry"</code></pre>
</div>
<p>label create the join.</p>
<div class="cell">
<pre><code>#&gt; Joining with `by = join_by(label)`</code></pre>
</div>
<p>Now we can visualize the first collapsed plot to make adjustments</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="c.-manual-improvement-of-collapsing-the-tree" class="level2">
<h2 class="anchored" data-anchor-id="c.-manual-improvement-of-collapsing-the-tree">C. Manual improvement of collapsing the tree</h2>
</section>
</section>
<section id="do-review" class="level1">
<h1>DO review</h1>
<ul>
<li>I need the node labels for review</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>branches where no isolates are included in Norway, it is more to be able to have overview of the position of our isolates within the tree.</li>
</ul>
<p>Selection from bottom to top.</p>
<p>Collapsing the additional nodes</p>
<div class="cell">
<pre><code>#&gt; recuring collapsing of nodes
#&gt; returning the collapsed tree</code></pre>
</div>
<p>Plotting the base of the collapsed tree</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Adding the metadata to the collapsed tree As previouslu</p>
<div class="cell">
<pre><code>#&gt; Joining with `by = join_by(label)`</code></pre>
</div>
<section id="d.-plotting-the-collapsed-tree" class="level2">
<h2 class="anchored" data-anchor-id="d.-plotting-the-collapsed-tree">D. Plotting the collapsed tree</h2>
<ul>
<li>first we make the base plot</li>
</ul>
<ul>
<li>Then we get the data for the annotations panels (it is required as separate dataframes)</li>
</ul>
<div class="cell">
<pre><code>#&gt;  [1] "parent"                     "node"                      
#&gt;  [3] "branch.length"              "label"                     
#&gt;  [5] "isTip"                      "x"                         
#&gt;  [7] "y"                          "branch"                    
#&gt;  [9] "angle"                      "labels_bootstrap"          
#&gt; [11] "node_depth"                 "descendants_tips_from_node"
#&gt; [13] "isMono_noboot"              "isMono"                    
#&gt; [15] "edge_num"                   "tip_metalink"              
#&gt; [17] "nested_country"             "ctg_country"               
#&gt; [19] "nested_Host"                "ctg_Host"                  
#&gt; [21] "nested_Host_latin"          "ctg_Host_latin"            
#&gt; [23] "nested_year"                "ctg_year"                  
#&gt; [25] "nested_continent"           "ctg_continent"             
#&gt; [27] "nested_Host_family"         "ctg_Host_family"           
#&gt; [29] "inNorway"                   "isWTE"                     
#&gt; [31] "category_highlight"         "isOneCountry"</code></pre>
</div>
<ol type="1">
<li>adding node that will be cut for detailed plotting with a light green point</li>
</ol>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Now we need to adjust the size of the panels</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><em>Figure caption:</em> H5N1 HA segment. ML phylogenetic tree with nodes depth &gt;= 5, that did not contained Norwegian isolates and nodes that represented groups of isolates from the same country of origin are collapsed. The tree is rooted, but the outgroup is not shown, due to large distance to the body of the tree. The information at the tips and collapsed nodes is representative of the data set, including identical sequences.</p>
</section>
</section>
<section id="components-for-final-visualisation" class="level1">
<h1>Components for final visualisation</h1>
<p>We will create the global circular tree And the two groups where White-tailed eagle are found to create trees that can be looked at more in detail.</p>
</section>
<section id="circular-tree" class="level1">
<h1>Circular tree</h1>
<p>We restart from the rooted tree with node labels</p>
<p>Circular metadata will be the at the last plots, except we do not add the coordinates</p>
<p>We had some plotting parameters to be able to view better</p>
<ul>
<li>Adding annotation around the tree</li>
</ul>
<div class="cell">
<pre><code>#&gt; Warning: Removed 25 rows containing missing values (`geom_point_g_gtree()`).</code></pre>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-60-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><em>Figure caption:</em> H5N1 HA segment. ML phylogenetic tree. The light green dots represent the branching origin the two groups presented as tree subsets. The tree is rooted, but the outgroup is not shown, due to large distance to the body of the tree. The information at the tips and collapsed nodes is representative of the data set, including identical sequences.”“”</p>
<div class="cell">
<pre><code>#&gt; Warning: Removed 25 rows containing missing values (`geom_point_g_gtree()`).</code></pre>
</div>
<section id="detailed-subplots" class="level2">
<h2 class="anchored" data-anchor-id="detailed-subplots">Detailed subplots</h2>
</section>
<section id="n348" class="level2">
<h2 class="anchored" data-anchor-id="n348">N348</h2>
<ul>
<li>getting the list of tips</li>
</ul>
<ul>
<li>subsetting the tree</li>
</ul>
<ul>
<li>creating the simple plot and adding the metadata</li>
</ul>
<ul>
<li>preparing the data for the pannels views</li>
</ul>
<ul>
<li>ploting the tree</li>
</ul>
<div class="cell">
<pre><code>#&gt; The min value of positional variable is 0.001200973 and the max value is 0.0144109585.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 2 sequences positions for your unnested variable
#&gt; The min value of positional variable is 0.001200973 and the max value is 0.0144109585.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 2 sequences positions for your unnested variable
#&gt; The min value of positional variable is 0.001200973 and the max value is 0.0144109585.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 2 sequences positions for your unnested variable
#&gt; The min value of positional variable is 0.001200973 and the max value is 0.0144109585.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 2 sequences positions for your unnested variable</code></pre>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-66-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Now we need to adjust the size of the panels</li>
</ul>
<div class="cell">
<pre><code>#&gt; The min value of positional variable is 0.001200973 and the max value is 0.0144109585.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 2 sequences positions for your unnested variable</code></pre>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-68-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><em>Figure Caption:</em> H5N1 HA segment. Detail of the second group of isolates from the ML phylogenetic tree. The subset tree is rooted and branches on the second green dot from the root. All the isolates are shown. Variables for sequences that were identical to the tip sequences are showned at the representative tips. The position of horizontal points is slighly adjusted to avoid overlap.”</p>
</section>
<section id="n9" class="level2">
<h2 class="anchored" data-anchor-id="n9">N9</h2>
<ul>
<li>getting the list of tips</li>
</ul>
<ul>
<li>subsetting the tree</li>
</ul>
<ul>
<li>creating the simple plot and adding the metadata</li>
</ul>
<ul>
<li>preparing the data for the pannels views</li>
</ul>
<ul>
<li>ploting the tree</li>
</ul>
<div class="cell">
<pre><code>#&gt; The min value of positional variable is 0.0011984215 and the max value is 0.013585641.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 2 sequences positions for your unnested variable
#&gt; The min value of positional variable is 0.0011984215 and the max value is 0.013585641.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 2 sequences positions for your unnested variable
#&gt; The min value of positional variable is 0.0011984215 and the max value is 0.013585641.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 2 sequences positions for your unnested variable
#&gt; The min value of positional variable is 0.0011984215 and the max value is 0.013585641.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 2 sequences positions for your unnested variable</code></pre>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-74-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Now we need to adjust the size of the panels</li>
</ul>
<div class="cell">
<pre><code>#&gt; The min value of positional variable is 0.0011984215 and the max value is 0.013585641.
#&gt;  
#&gt;                   This should help you choose spacing values
#&gt; You have a maximum of 2 sequences positions for your unnested variable</code></pre>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="SEP_H5N1_HA_phylogenetic_tree_files/figure-html/unnamed-chunk-76-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><em>Figure Caption:</em></p>
<p>H5N1 HA segment. Detail of the first group of isolates from the ML phylogenetic tree. The subset tree is rooted and branches on the second green dot from the root. All the isolates are shown. Variables for sequences that were identical to the tip sequences are showned at the representative tips. The position of horizontal points is slighly adjusted to avoid overlap.”</p>
</section>
</section>
<section id="saving-environment" class="level1">
<h1>Saving environment</h1>
<p>To continue analyses that are not finished</p>
<p>List of isolates used in the tree (otherwise is in the image)</p>
<p>Metadata for all H5N1 isolates included</p>
<p>This dataset is representative for 2378 isolates. This includes 1421 isolates with unique sequences, including 2 sequences as ougroup.</p>
<p>Note that one group of isolate isolates belonging to the same year of sampling has been removed from our data set in an attempt to reduce the size of the tree. This after a preliminary analysis of the data. This group of isolates appeared to originate from our second subset.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="%22results/H5N1_HA/complete.svg%22" class="img-fluid figure-img"></p>
<figcaption>Complete tree, preliminary analysis</figcaption>
</figure>
</div>
</section>
<section id="notes" class="level1">
<h1>Notes</h1>
<p>Data control is ok - its only H5N1 <!-- I only had done my join in the wrong way --></p>
<p>Some isolates have been submitted to GISAID since we started this analysis, therefore the complement of GISAID_ID will be completed manually.</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>